// Data.Monoid.All/index.js
// Generated by psc-make version 0.6.9.5
"use strict";
var Prelude = require("Prelude");
var Data_Monoid = require("Data.Monoid");
var All = function (x) {
    return x;
};
var showAll = new Prelude.Show(function (_182) {
    return "All (" + (Prelude.show(Prelude.showBoolean)(_182) + ")");
});
var semigroupAll = new Prelude.Semigroup(function (_183) {
    return function (_184) {
        return _183 && _184;
    };
});
var runAll = function (_175) {
    return _175;
};
var monoidAll = new Data_Monoid.Monoid(function () {
    return semigroupAll;
}, true);
var eqAll = new Prelude.Eq(function (_178) {
    return function (_179) {
        return _178 !== _179;
    };
}, function (_176) {
    return function (_177) {
        return _176 === _177;
    };
});
var ordAll = new Prelude.Ord(function () {
    return eqAll;
}, function (_180) {
    return function (_181) {
        return Prelude.compare(Prelude.ordBoolean)(_180)(_181);
    };
});
module.exports = {
    All: All, 
    runAll: runAll, 
    eqAll: eqAll, 
    ordAll: ordAll, 
    showAll: showAll, 
    semigroupAll: semigroupAll, 
    monoidAll: monoidAll
};
