// Data.Monoid.All/index.js
// Generated by psc-make version 0.6.9.5
"use strict";
var Prelude = require("Prelude");
var Data_Monoid = require("Data.Monoid");
var All = function (x) {
    return x;
};
var showAll = new Prelude.Show(function (_191) {
    return "All (" + (Prelude.show(Prelude.showBoolean)(_191) + ")");
});
var semigroupAll = new Prelude.Semigroup(function (_192) {
    return function (_193) {
        return _192 && _193;
    };
});
var runAll = function (_184) {
    return _184;
};
var monoidAll = new Data_Monoid.Monoid(function () {
    return semigroupAll;
}, true);
var eqAll = new Prelude.Eq(function (_187) {
    return function (_188) {
        return _187 !== _188;
    };
}, function (_185) {
    return function (_186) {
        return _185 === _186;
    };
});
var ordAll = new Prelude.Ord(function () {
    return eqAll;
}, function (_189) {
    return function (_190) {
        return Prelude.compare(Prelude.ordBoolean)(_189)(_190);
    };
});
module.exports = {
    All: All, 
    runAll: runAll, 
    eqAll: eqAll, 
    ordAll: ordAll, 
    showAll: showAll, 
    semigroupAll: semigroupAll, 
    monoidAll: monoidAll
};
