// Data.Monoid.All/index.js
// Generated by psc-make version 0.6.9.5
"use strict";
var Prelude = require("Prelude");
var Data_Monoid = require("Data.Monoid");
var All = function (x) {
    return x;
};
var showAll = new Prelude.Show(function (_171) {
    return "All (" + (Prelude.show(Prelude.showBoolean)(_171) + ")");
});
var semigroupAll = new Prelude.Semigroup(function (_172) {
    return function (_173) {
        return _172 && _173;
    };
});
var runAll = function (_164) {
    return _164;
};
var monoidAll = new Data_Monoid.Monoid(function () {
    return semigroupAll;
}, true);
var eqAll = new Prelude.Eq(function (_167) {
    return function (_168) {
        return _167 !== _168;
    };
}, function (_165) {
    return function (_166) {
        return _165 === _166;
    };
});
var ordAll = new Prelude.Ord(function () {
    return eqAll;
}, function (_169) {
    return function (_170) {
        return Prelude.compare(Prelude.ordBoolean)(_169)(_170);
    };
});
module.exports = {
    All: All, 
    runAll: runAll, 
    eqAll: eqAll, 
    ordAll: ordAll, 
    showAll: showAll, 
    semigroupAll: semigroupAll, 
    monoidAll: monoidAll
};
