// Data.Char/index.js
// Generated by psc-make version 0.6.9.5
"use strict";
var Prelude = require("Prelude");

    function toCharCode(c) {
      return c.charCodeAt(0);
    }
    ;

    function fromCharCode(c) {
      return String.fromCharCode(c);
    }
    ;
var Char = function (x) {
    return x;
};
var showChar = new Prelude.Show(function (_85) {
    return "Char " + Prelude.show(Prelude.showString)(_85);
});
var eqChar = new Prelude.Eq(function (a) {
    return function (b) {
        return !Prelude["=="](eqChar)(a)(b);
    };
}, function (_81) {
    return function (_82) {
        return _81 === _82;
    };
});
var ordChar = new Prelude.Ord(function () {
    return eqChar;
}, function (_83) {
    return function (_84) {
        return Prelude.compare(Prelude.ordString)(_83)(_84);
    };
});
var charString = function (_80) {
    return _80;
};
module.exports = {
    toCharCode: toCharCode, 
    fromCharCode: fromCharCode, 
    charString: charString, 
    eqChar: eqChar, 
    ordChar: ordChar, 
    showChar: showChar
};
