// Control.Monad.Aff.Par/index.js
// Generated by psc-make version 0.6.9.5
"use strict";
var Prelude = require("Prelude");
var Data_Monoid = require("Data.Monoid");
var Control_Monad_Aff_AVar = require("Control.Monad.Aff.AVar");
var Control_Monad_Aff = require("Control.Monad.Aff");
var Data_Either = require("Data.Either");
var Control_Plus = require("Control.Plus");
var Control_Apply = require("Control.Apply");
var Control_Alt = require("Control.Alt");
var Control_Alternative = require("Control.Alternative");
var Control_Monad_Error_Class = require("Control.Monad.Error.Class");
var Par = function (x) {
    return x;
};
var runPar = function (_298) {
    return _298;
};
var functorPar = new Prelude.Functor(function (f) {
    return function (_299) {
        return Prelude["<$>"](Control_Monad_Aff.functorAff)(f)(_299);
    };
});
var applyPar = new Prelude.Apply(function (_300) {
    return function (_301) {
        return Prelude[">>="](Control_Monad_Aff.bindAff)(Control_Monad_Aff_AVar.makeVar)(function (_32) {
            return Prelude[">>="](Control_Monad_Aff.bindAff)(Control_Monad_Aff_AVar.makeVar)(function (_31) {
                return Prelude[">>="](Control_Monad_Aff.bindAff)(Control_Monad_Aff.forkAff(Prelude[">>="](Control_Monad_Aff.bindAff)(_300)(Control_Monad_Aff_AVar.putVar(_32))))(function (_30) {
                    return Prelude[">>="](Control_Monad_Aff.bindAff)(Control_Monad_Aff.forkAff(Prelude[">>="](Control_Monad_Aff.bindAff)(_301)(Control_Monad_Aff_AVar.putVar(_31))))(function (_29) {
                        return Control_Monad_Aff.cancelWith(Prelude["<*>"](Control_Monad_Aff.applyAff)(Control_Monad_Aff_AVar.takeVar(_32))(Control_Monad_Aff_AVar.takeVar(_31)))(Prelude["<>"](Control_Monad_Aff.semigroupCanceler)(_30)(_29));
                    });
                });
            });
        });
    };
}, function () {
    return functorPar;
});
var semigroupPar = function (__dict_Semigroup_0) {
    return new Prelude.Semigroup(function (a) {
        return function (b) {
            return Prelude["<*>"](applyPar)(Prelude["<$>"](functorPar)(Prelude["<>"](__dict_Semigroup_0))(a))(b);
        };
    });
};
var applicativePar = new Prelude.Applicative(function () {
    return applyPar;
}, function (v) {
    return Prelude.pure(Control_Monad_Aff.applicativeAff)(v);
});
var monoidPar = function (__dict_Monoid_1) {
    return new Data_Monoid.Monoid(function () {
        return semigroupPar(__dict_Monoid_1["__superclass_Prelude.Semigroup_0"]());
    }, Prelude.pure(applicativePar)(Data_Monoid.mempty(__dict_Monoid_1)));
};
var altPar = new Control_Alt.Alt(function (_302) {
    return function (_303) {
        var maybeKill = function (va) {
            return function (ve) {
                return function (err) {
                    return Prelude[">>="](Control_Monad_Aff.bindAff)(Control_Monad_Aff_AVar.takeVar(ve))(function (_33) {
                        return Prelude[">>="](Control_Monad_Aff.bindAff)((function () {
                            var _1281 = _33 === 1;
                            if (_1281) {
                                return Control_Monad_Aff_AVar.killVar(va)(err);
                            };
                            if (!_1281) {
                                return Prelude["return"](Control_Monad_Aff.monadAff)(Prelude.unit);
                            };
                            throw new Error("Failed pattern match");
                        })())(function () {
                            return Control_Monad_Aff_AVar.putVar(ve)(_33 + 1);
                        });
                    });
                };
            };
        };
        return Prelude[">>="](Control_Monad_Aff.bindAff)(Control_Monad_Aff_AVar.makeVar)(function (_37) {
            return Prelude[">>="](Control_Monad_Aff.bindAff)(Control_Monad_Aff_AVar["makeVar'"](0))(function (_36) {
                return Prelude[">>="](Control_Monad_Aff.bindAff)(Control_Monad_Aff.forkAff(Prelude[">>="](Control_Monad_Aff.bindAff)(Control_Monad_Aff.attempt(_302))(Data_Either.either(maybeKill(_37)(_36))(Control_Monad_Aff_AVar.putVar(_37)))))(function (_35) {
                    return Prelude[">>="](Control_Monad_Aff.bindAff)(Control_Monad_Aff.forkAff(Prelude[">>="](Control_Monad_Aff.bindAff)(Control_Monad_Aff.attempt(_303))(Data_Either.either(maybeKill(_37)(_36))(Control_Monad_Aff_AVar.putVar(_37)))))(function (_34) {
                        return Control_Monad_Aff.cancelWith(Control_Monad_Aff_AVar.takeVar(_37))(Prelude["<>"](Control_Monad_Aff.semigroupCanceler)(_35)(_34));
                    });
                });
            });
        });
    };
}, function () {
    return functorPar;
});
var plusPar = new Control_Plus.Plus(function () {
    return altPar;
}, Control_Plus.empty(Control_Monad_Aff.plusAff));
var alternativePar = new Control_Alternative.Alternative(function () {
    return plusPar;
}, function () {
    return applicativePar;
});
module.exports = {
    Par: Par, 
    runPar: runPar, 
    semigroupPar: semigroupPar, 
    monoidPar: monoidPar, 
    functorPar: functorPar, 
    applyPar: applyPar, 
    applicativePar: applicativePar, 
    altPar: altPar, 
    plusPar: plusPar, 
    alternativePar: alternativePar
};
