// Data.Sequence.NonEmpty/index.js
// Generated by psc-make version 0.6.9.5
"use strict";
var Data_Sequence = require("Data.Sequence");
var Prelude = require("Prelude");
var Data_Tuple = require("Data.Tuple");
var Data_Maybe = require("Data.Maybe");
var Data_Foldable = require("Data.Foldable");
var Data_Traversable = require("Data.Traversable");
var Control_Alt = require("Control.Alt");
var Data_Unfoldable = require("Data.Unfoldable");
var Seq = (function () {
    function Seq(value0, value1) {
        this.value0 = value0;
        this.value1 = value1;
    };
    Seq.create = function (value0) {
        return function (value1) {
            return new Seq(value0, value1);
        };
    };
    return Seq;
})();
var unsnoc = function (_679) {
    var _3142 = Data_Sequence.unsnoc(_679.value1);
    if (_3142 instanceof Data_Maybe.Nothing) {
        return new Data_Tuple.Tuple(Data_Sequence.empty, _679.value0);
    };
    if (_3142 instanceof Data_Maybe.Just) {
        return new Data_Tuple.Tuple(Data_Sequence.cons(_679.value0)(_3142.value0.value0), _3142.value0.value1);
    };
    throw new Error("Failed pattern match");
};
var uncons = function (_678) {
    return new Data_Tuple.Tuple(_678.value0, _678.value1);
};
var toPlain = function (_683) {
    return Data_Sequence.cons(_683.value0)(_683.value1);
};
var take = function (i) {
    return Prelude["<<<"](Prelude.semigroupoidArr)(Data_Sequence.take(i))(toPlain);
};
var tail = function (_681) {
    return _681.value1;
};
var splitAt = function (i) {
    return Prelude["<<<"](Prelude.semigroupoidArr)(Data_Sequence.splitAt(i))(toPlain);
};
var snoc = function (_672) {
    return function (y) {
        return new Seq(_672.value0, Data_Sequence.snoc(_672.value1)(y));
    };
};
var singleton = function (x) {
    return new Seq(x, Data_Sequence.empty);
};
var showSeq = function (__dict_Show_0) {
    return new Prelude.Show(function (_688) {
        return "(Seq " + (Prelude.show(__dict_Show_0)(_688.value0) + (" " + (Prelude.show(Data_Sequence.showSeq(__dict_Show_0))(_688.value1) + ")")));
    });
};
var semigroupSeq = new Prelude.Semigroup(function (_694) {
    return function (_695) {
        return new Seq(_694.value0, Prelude["<>"](Data_Sequence.semigroupSeq)(Data_Sequence.snoc(_694.value1)(_695.value0))(_695.value1));
    };
});
var length = function (_675) {
    return Data_Sequence.length(_675.value1) + 1;
};
var last = function (_682) {
    return Data_Maybe.maybe(_682.value0)(Prelude.id(Prelude.categoryArr))(Data_Sequence.last(_682.value1));
};
var init = Prelude["<<<"](Prelude.semigroupoidArr)(Data_Tuple.fst)(unsnoc);
var index = function (_684) {
    return function (_685) {
        if (_685 === 0) {
            return new Data_Maybe.Just(_684.value0);
        };
        return Data_Sequence.index(_684.value1)(_685 - 1);
    };
};
var inBounds = function (_676) {
    return function (_677) {
        if (_677 === 0) {
            return true;
        };
        return Data_Sequence.inBounds(_676.value1)(_677 - 1);
    };
};
var head = function (_680) {
    return _680.value0;
};
var functorSeq = new Prelude.Functor(function (f) {
    return function (_693) {
        return new Seq(f(_693.value0), Prelude["<$>"](Data_Sequence.functorSeq)(f)(_693.value1));
    };
});
var fromSeq = function (__dict_Functor_2) {
    return function (__dict_Unfoldable_3) {
        return Prelude["<<<"](Prelude.semigroupoidArr)(Data_Sequence.fromSeq(__dict_Functor_2)(__dict_Unfoldable_3))(toPlain);
    };
};
var fromPlainUnsafe = function (xs) {
    var _3193 = Data_Sequence.uncons(xs);
    if (_3193 instanceof Data_Maybe.Just) {
        return new Seq(_3193.value0.value0, _3193.value0.value1);
    };
    throw new Error("Failed pattern match");
};
var foldableSeq = new Data_Foldable.Foldable(function (__dict_Monoid_4) {
    return function (f) {
        return Prelude[">>>"](Prelude.semigroupoidArr)(toPlain)(Data_Foldable.foldMap(Data_Sequence.foldableSeq)(__dict_Monoid_4)(f));
    };
}, function (f) {
    return function (z) {
        return Prelude[">>>"](Prelude.semigroupoidArr)(toPlain)(Data_Foldable.foldl(Data_Sequence.foldableSeq)(f)(z));
    };
}, function (f) {
    return function (z) {
        return Prelude[">>>"](Prelude.semigroupoidArr)(toPlain)(Data_Foldable.foldr(Data_Sequence.foldableSeq)(f)(z));
    };
});
var fmap = Prelude["<$>"];
var traversableSeq = new Data_Traversable.Traversable(function () {
    return foldableSeq;
}, function () {
    return functorSeq;
}, function (__dict_Applicative_5) {
    return Prelude[">>>"](Prelude.semigroupoidArr)(toPlain)(Prelude[">>>"](Prelude.semigroupoidArr)(Data_Traversable.sequence(Data_Sequence.traversableSeq)(__dict_Applicative_5))(fmap((__dict_Applicative_5["__superclass_Prelude.Apply_0"]())["__superclass_Prelude.Functor_0"]())(fromPlainUnsafe)));
}, function (__dict_Applicative_6) {
    return function (f) {
        return Prelude[">>>"](Prelude.semigroupoidArr)(toPlain)(Prelude[">>>"](Prelude.semigroupoidArr)(Data_Traversable.traverse(Data_Sequence.traversableSeq)(__dict_Applicative_6)(f))(fmap((__dict_Applicative_6["__superclass_Prelude.Apply_0"]())["__superclass_Prelude.Functor_0"]())(fromPlainUnsafe)));
    };
});
var filter = function (p) {
    return Prelude["<<<"](Prelude.semigroupoidArr)(Data_Sequence.filter(p))(toPlain);
};
var eqSeq = function (__dict_Eq_7) {
    return new Prelude.Eq(function (x) {
        return function (y) {
            return !Prelude["=="](eqSeq(__dict_Eq_7))(x)(y);
        };
    }, function (_689) {
        return function (_690) {
            return Prelude["=="](__dict_Eq_7)(_689.value0)(_690.value0) && Prelude["=="](Data_Sequence.eqSeq(__dict_Eq_7))(_689.value1)(_690.value1);
        };
    });
};
var ordSeq = function (__dict_Ord_1) {
    return new Prelude.Ord(function () {
        return eqSeq(__dict_Ord_1["__superclass_Prelude.Eq_0"]());
    }, function (_691) {
        return function (_692) {
            var _3205 = Prelude.compare(__dict_Ord_1)(_691.value0)(_692.value0);
            if (_3205 instanceof Prelude.EQ) {
                return Prelude.compare(Data_Sequence.ordSeq(__dict_Ord_1))(_691.value1)(_692.value1);
            };
            return _3205;
        };
    });
};
var drop = function (i) {
    return Prelude["<<<"](Prelude.semigroupoidArr)(Data_Sequence.drop(i))(toPlain);
};
var cons = function (x) {
    return function (_671) {
        return new Seq(x, Data_Sequence.cons(_671.value0)(_671.value1));
    };
};
var applySeq = new Prelude.Apply(function (fs) {
    return function (xs) {
        return fromPlainUnsafe(Prelude["<*>"](Data_Sequence.applySeq)(toPlain(fs))(toPlain(xs)));
    };
}, function () {
    return functorSeq;
});
var bindSeq = new Prelude.Bind(function (xs) {
    return function (f) {
        return fromPlainUnsafe(Prelude[">>="](Data_Sequence.bindSeq)(toPlain(xs))(Prelude["<<<"](Prelude.semigroupoidArr)(toPlain)(f)));
    };
}, function () {
    return applySeq;
});
var applicativeSeq = new Prelude.Applicative(function () {
    return applySeq;
}, function (x) {
    return new Seq(x, Data_Sequence.empty);
});
var monadSeq = new Prelude.Monad(function () {
    return applicativeSeq;
}, function () {
    return bindSeq;
});
var append = function (_673) {
    return function (_674) {
        return new Seq(_673.value0, Prelude["<>"](Data_Sequence.semigroupSeq)(Data_Sequence.snoc(_673.value1)(_674.value0))(_674.value1));
    };
};
var altSeq = new Control_Alt.Alt(Prelude["<>"](semigroupSeq), function () {
    return functorSeq;
});
var adjust = function (f) {
    return function (_686) {
        return function (_687) {
            if (_686 === 0) {
                return new Seq(f(_687.value0), _687.value1);
            };
            return new Seq(_687.value0, Data_Sequence.adjust(f)(_686 - 1)(_687.value1));
        };
    };
};
var replace = function (x) {
    return adjust(Prelude["const"](x));
};
module.exports = {
    Seq: Seq, 
    fromSeq: fromSeq, 
    replace: replace, 
    adjust: adjust, 
    index: index, 
    filter: filter, 
    drop: drop, 
    take: take, 
    splitAt: splitAt, 
    toPlain: toPlain, 
    last: last, 
    init: init, 
    tail: tail, 
    head: head, 
    unsnoc: unsnoc, 
    uncons: uncons, 
    inBounds: inBounds, 
    length: length, 
    append: append, 
    snoc: snoc, 
    cons: cons, 
    singleton: singleton, 
    showSeq: showSeq, 
    eqSeq: eqSeq, 
    ordSeq: ordSeq, 
    functorSeq: functorSeq, 
    applySeq: applySeq, 
    applicativeSeq: applicativeSeq, 
    bindSeq: bindSeq, 
    monadSeq: monadSeq, 
    semigroupSeq: semigroupSeq, 
    altSeq: altSeq, 
    foldableSeq: foldableSeq, 
    traversableSeq: traversableSeq
};
