// Control.Monad.Cont.Class/index.js
// Generated by psc-make version 0.6.9.5
"use strict";
var Control_Monad_Cont_Trans = require("Control.Monad.Cont.Trans");
var Control_Monad_Error_Trans = require("Control.Monad.Error.Trans");
var Control_Monad_Maybe_Trans = require("Control.Monad.Maybe.Trans");
var Control_Monad_Reader_Trans = require("Control.Monad.Reader.Trans");
var Control_Monad_State_Trans = require("Control.Monad.State.Trans");
var Control_Monad_Writer_Trans = require("Control.Monad.Writer.Trans");
var Prelude = require("Prelude");
var Control_Monad_Error = require("Control.Monad.Error");
var Data_Monoid = require("Data.Monoid");
var MonadCont = function (callCC) {
    this.callCC = callCC;
};
var monadContContT = function (__dict_Monad_0) {
    return new MonadCont(Control_Monad_Cont_Trans.callCC);
};
var callCC = function (dict) {
    return dict.callCC;
};
var monadContErrorT = function (__dict_Error_1) {
    return function (__dict_MonadCont_2) {
        return new MonadCont(Control_Monad_Error_Trans.liftCallCCError(callCC(__dict_MonadCont_2)));
    };
};
var monadContMaybeT = function (__dict_MonadCont_3) {
    return new MonadCont(Control_Monad_Maybe_Trans.liftCallCCMaybe(callCC(__dict_MonadCont_3)));
};
var monadContReaderT = function (__dict_MonadCont_4) {
    return new MonadCont(Control_Monad_Reader_Trans.liftCallCCReader(callCC(__dict_MonadCont_4)));
};
var monadContStateT = function (__dict_MonadCont_5) {
    return new MonadCont(Control_Monad_State_Trans["liftCallCCState'"](callCC(__dict_MonadCont_5)));
};
var monadWriterT = function (__dict_Monoid_6) {
    return function (__dict_MonadCont_7) {
        return new MonadCont(Control_Monad_Writer_Trans.liftCallCCWriter(__dict_Monoid_6)(callCC(__dict_MonadCont_7)));
    };
};
module.exports = {
    MonadCont: MonadCont, 
    callCC: callCC, 
    monadContContT: monadContContT, 
    monadContErrorT: monadContErrorT, 
    monadContMaybeT: monadContMaybeT, 
    monadContReaderT: monadContReaderT, 
    monadContStateT: monadContStateT, 
    monadWriterT: monadWriterT
};
