// Data.FingerTree/index.js
// Generated by psc-make version 0.6.9.5
"use strict";
var Prelude = require("Prelude");
var Data_Sequence_Internal = require("Data.Sequence.Internal");
var Data_Foldable = require("Data.Foldable");
var Data_Monoid = require("Data.Monoid");
var Data_Traversable = require("Data.Traversable");
var Data_Lazy = require("Data.Lazy");
var Data_Array = require("Data.Array");
var Data_Array_Unsafe = require("Data.Array.Unsafe");
var Data_Unfoldable = require("Data.Unfoldable");
var Data_Maybe = require("Data.Maybe");
var Data_Tuple = require("Data.Tuple");
var NilR = (function () {
    function NilR() {

    };
    NilR.value = new NilR();
    return NilR;
})();
var SnocR = (function () {
    function SnocR(value0, value1) {
        this.value0 = value0;
        this.value1 = value1;
    };
    SnocR.create = function (value0) {
        return function (value1) {
            return new SnocR(value0, value1);
        };
    };
    return SnocR;
})();
var NilL = (function () {
    function NilL() {

    };
    NilL.value = new NilL();
    return NilL;
})();
var ConsL = (function () {
    function ConsL(value0, value1) {
        this.value0 = value0;
        this.value1 = value1;
    };
    ConsL.create = function (value0) {
        return function (value1) {
            return new ConsL(value0, value1);
        };
    };
    return ConsL;
})();
var Split = (function () {
    function Split(value0, value1, value2) {
        this.value0 = value0;
        this.value1 = value1;
        this.value2 = value2;
    };
    Split.create = function (value0) {
        return function (value1) {
            return function (value2) {
                return new Split(value0, value1, value2);
            };
        };
    };
    return Split;
})();
var Node2 = (function () {
    function Node2(value0, value1, value2) {
        this.value0 = value0;
        this.value1 = value1;
        this.value2 = value2;
    };
    Node2.create = function (value0) {
        return function (value1) {
            return function (value2) {
                return new Node2(value0, value1, value2);
            };
        };
    };
    return Node2;
})();
var Node3 = (function () {
    function Node3(value0, value1, value2, value3) {
        this.value0 = value0;
        this.value1 = value1;
        this.value2 = value2;
        this.value3 = value3;
    };
    Node3.create = function (value0) {
        return function (value1) {
            return function (value2) {
                return function (value3) {
                    return new Node3(value0, value1, value2, value3);
                };
            };
        };
    };
    return Node3;
})();
var LazySplit = (function () {
    function LazySplit(value0, value1, value2) {
        this.value0 = value0;
        this.value1 = value1;
        this.value2 = value2;
    };
    LazySplit.create = function (value0) {
        return function (value1) {
            return function (value2) {
                return new LazySplit(value0, value1, value2);
            };
        };
    };
    return LazySplit;
})();
var Empty = (function () {
    function Empty() {

    };
    Empty.value = new Empty();
    return Empty;
})();
var Single = (function () {
    function Single(value0) {
        this.value0 = value0;
    };
    Single.create = function (value0) {
        return new Single(value0);
    };
    return Single;
})();
var Deep = (function () {
    function Deep(value0, value1, value2, value3) {
        this.value0 = value0;
        this.value1 = value1;
        this.value2 = value2;
        this.value3 = value3;
    };
    Deep.create = function (value0) {
        return function (value1) {
            return function (value2) {
                return function (value3) {
                    return new Deep(value0, value1, value2, value3);
                };
            };
        };
    };
    return Deep;
})();
var tailDigit = Data_Array_Unsafe.tail;
var splitDigit = function (__dict_Monoid_4) {
    return function (__dict_Measured_5) {
        return function (p) {
            return function (i) {
                return function (as) {
                    var _2340 = Data_Array.length(as);
                    if (_2340 === 1) {
                        return new Split([  ], Data_Sequence_Internal["!"](as)(0), [  ]);
                    };
                    var bs = Data_Array.drop(1)(as);
                    var a = Data_Sequence_Internal["!"](as)(0);
                    var i$prime = Prelude["<>"](__dict_Monoid_4["__superclass_Prelude.Semigroup_0"]())(i)(Data_Sequence_Internal.measure(__dict_Measured_5)(a));
                    var _2341 = p(i$prime);
                    if (_2341) {
                        return new Split([  ], a, bs);
                    };
                    if (!_2341) {
                        var _2342 = splitDigit(__dict_Monoid_4)(__dict_Measured_5)(p)(i$prime)(bs);
                        return new Split(Prelude[":"](a)(_2342.value0), _2342.value1, _2342.value2);
                    };
                    throw new Error("Failed pattern match");
                };
            };
        };
    };
};
var showNode = function (__dict_Show_6) {
    return function (__dict_Show_7) {
        return new Prelude.Show(function (_518) {
            if (_518 instanceof Node2) {
                return "Node2 (" + (Prelude.show(__dict_Show_7)(_518.value0) + (") (" + (Prelude.show(__dict_Show_6)(_518.value1) + (") (" + (Prelude.show(__dict_Show_6)(_518.value2) + ")")))));
            };
            if (_518 instanceof Node3) {
                return "Node3 (" + (Prelude.show(__dict_Show_7)(_518.value0) + (") (" + (Prelude.show(__dict_Show_6)(_518.value1) + (") (" + (Prelude.show(__dict_Show_6)(_518.value2) + (") (" + (Prelude.show(__dict_Show_6)(_518.value3) + ")")))))));
            };
            throw new Error("Failed pattern match");
        });
    };
};
var showFingerTree = function (__dict_Show_8) {
    return function (__dict_Show_9) {
        return new Prelude.Show(function (_524) {
            if (_524 instanceof Empty) {
                return "Empty";
            };
            if (_524 instanceof Single) {
                return "Single (" + (Prelude.show(__dict_Show_9)(_524.value0) + ")");
            };
            if (_524 instanceof Deep) {
                return "Deep (" + (Prelude.show(Data_Lazy.showLazy(__dict_Show_8))(_524.value0) + (") (" + (Prelude.show(Prelude.showArray(__dict_Show_9))(_524.value1) + (") (" + (Prelude.show(Data_Lazy.showLazy(showFingerTree(__dict_Show_8)(showNode(__dict_Show_9)(__dict_Show_8))))(_524.value2) + (") (" + (Prelude.show(Prelude.showArray(__dict_Show_9))(_524.value3) + ")")))))));
            };
            throw new Error("Failed pattern match");
        });
    };
};
var nodeToDigit = function (_506) {
    if (_506 instanceof Node2) {
        return [ _506.value1, _506.value2 ];
    };
    if (_506 instanceof Node3) {
        return [ _506.value1, _506.value2, _506.value3 ];
    };
    throw new Error("Failed pattern match");
};
var node3 = function (__dict_Monoid_10) {
    return function (__dict_Measured_11) {
        return function (a) {
            return function (b) {
                return function (c) {
                    return new Node3(Prelude["<>"](__dict_Monoid_10["__superclass_Prelude.Semigroup_0"]())(Data_Sequence_Internal.measure(__dict_Measured_11)(a))(Prelude["<>"](__dict_Monoid_10["__superclass_Prelude.Semigroup_0"]())(Data_Sequence_Internal.measure(__dict_Measured_11)(b))(Data_Sequence_Internal.measure(__dict_Measured_11)(c))), a, b, c);
                };
            };
        };
    };
};
var node2 = function (__dict_Monoid_12) {
    return function (__dict_Measured_13) {
        return function (a) {
            return function (b) {
                return new Node2(Prelude["<>"](__dict_Monoid_12["__superclass_Prelude.Semigroup_0"]())(Data_Sequence_Internal.measure(__dict_Measured_13)(a))(Data_Sequence_Internal.measure(__dict_Measured_13)(b)), a, b);
            };
        };
    };
};
var nodes = function (__dict_Monoid_14) {
    return function (__dict_Measured_15) {
        return function (_515) {
            if (_515.length === 2) {
                return [ node2(__dict_Monoid_14)(__dict_Measured_15)(_515[0])(_515[1]) ];
            };
            if (_515.length === 3) {
                return [ node3(__dict_Monoid_14)(__dict_Measured_15)(_515[0])(_515[1])(_515[2]) ];
            };
            if (_515.length === 4) {
                return [ node2(__dict_Monoid_14)(__dict_Measured_15)(_515[0])(_515[1]), node2(__dict_Monoid_14)(__dict_Measured_15)(_515[2])(_515[3]) ];
            };
            return Prelude[":"](node3(__dict_Monoid_14)(__dict_Measured_15)(Data_Sequence_Internal["!"](_515)(0))(Data_Sequence_Internal["!"](_515)(1))(Data_Sequence_Internal["!"](_515)(2)))(nodes(__dict_Monoid_14)(__dict_Measured_15)(Data_Array.drop(3)(_515)));
        };
    };
};
var measuredNode = new Data_Sequence_Internal.Measured(function (_523) {
    if (_523 instanceof Node2) {
        return _523.value0;
    };
    if (_523 instanceof Node3) {
        return _523.value0;
    };
    throw new Error("Failed pattern match");
});
var measuredFingerTree = function (__dict_Monoid_16) {
    return function (__dict_Measured_17) {
        return new Data_Sequence_Internal.Measured(function (_529) {
            if (_529 instanceof Empty) {
                return Data_Monoid.mempty(__dict_Monoid_16);
            };
            if (_529 instanceof Single) {
                return Data_Sequence_Internal.measure(__dict_Measured_17)(_529.value0);
            };
            if (_529 instanceof Deep) {
                return Data_Lazy.force(_529.value0);
            };
            throw new Error("Failed pattern match");
        });
    };
};
var lazyEmpty = Data_Lazy.defer(function (_492) {
    return Empty.value;
});
var lastDigit = Data_Array_Unsafe.last;
var initDigit = Data_Array_Unsafe.init;
var headDigit = Data_Array_Unsafe.head;
var functorViewL = function (__dict_Functor_18) {
    return new Prelude.Functor(function (f) {
        return function (_530) {
            if (_530 instanceof NilL) {
                return NilL.value;
            };
            if (_530 instanceof ConsL) {
                return new ConsL(f(_530.value0), Prelude["<$>"](Data_Lazy.functorLazy)(Prelude["<$>"](__dict_Functor_18)(f))(_530.value1));
            };
            throw new Error("Failed pattern match");
        };
    });
};
var functorNode = new Prelude.Functor(function (f) {
    return function (_519) {
        if (_519 instanceof Node2) {
            return new Node2(_519.value0, f(_519.value1), f(_519.value2));
        };
        if (_519 instanceof Node3) {
            return new Node3(_519.value0, f(_519.value1), f(_519.value2), f(_519.value3));
        };
        throw new Error("Failed pattern match");
    };
});
var functorFingerTree = new Prelude.Functor(function (f) {
    return function (_525) {
        if (_525 instanceof Empty) {
            return Empty.value;
        };
        if (_525 instanceof Single) {
            return new Single(f(_525.value0));
        };
        if (_525 instanceof Deep) {
            return new Deep(_525.value0, Prelude["<$>"](Data_Array.functorArray)(f)(_525.value1), Data_Sequence_Internal["<$$$>"](Data_Lazy.functorLazy)(functorFingerTree)(functorNode)(f)(_525.value2), Prelude["<$>"](Data_Array.functorArray)(f)(_525.value3));
        };
        throw new Error("Failed pattern match");
    };
});
var foldableNode = new Data_Foldable.Foldable(function (__dict_Monoid_19) {
    return function (f) {
        return function (xs) {
            return Data_Foldable.foldr(foldableNode)(function (x) {
                return function (acc) {
                    return Prelude["<>"](__dict_Monoid_19["__superclass_Prelude.Semigroup_0"]())(f(x))(acc);
                };
            })(Data_Monoid.mempty(__dict_Monoid_19))(xs);
        };
    };
}, function ($greater$minus) {
    return function (z) {
        return function (_521) {
            if (_521 instanceof Node2) {
                return $greater$minus($greater$minus(z)(_521.value1))(_521.value2);
            };
            if (_521 instanceof Node3) {
                return $greater$minus($greater$minus($greater$minus(z)(_521.value1))(_521.value2))(_521.value3);
            };
            throw new Error("Failed pattern match");
        };
    };
}, function ($minus$less) {
    return function (z) {
        return function (_520) {
            if (_520 instanceof Node2) {
                return $minus$less(_520.value1)($minus$less(_520.value2)(z));
            };
            if (_520 instanceof Node3) {
                return $minus$less(_520.value1)($minus$less(_520.value2)($minus$less(_520.value3)(z)));
            };
            throw new Error("Failed pattern match");
        };
    };
});
var traversableNode = new Data_Traversable.Traversable(function () {
    return foldableNode;
}, function () {
    return functorNode;
}, function (__dict_Applicative_1) {
    return Data_Traversable.traverse(traversableNode)(__dict_Applicative_1)(Prelude.id(Prelude.categoryArr));
}, function (__dict_Applicative_0) {
    return function (f) {
        return function (_522) {
            if (_522 instanceof Node2) {
                return Prelude["<*>"](__dict_Applicative_0["__superclass_Prelude.Apply_0"]())(Prelude["<$>"]((__dict_Applicative_0["__superclass_Prelude.Apply_0"]())["__superclass_Prelude.Functor_0"]())(Node2.create(_522.value0))(f(_522.value1)))(f(_522.value2));
            };
            if (_522 instanceof Node3) {
                return Prelude["<*>"](__dict_Applicative_0["__superclass_Prelude.Apply_0"]())(Prelude["<*>"](__dict_Applicative_0["__superclass_Prelude.Apply_0"]())(Prelude["<$>"]((__dict_Applicative_0["__superclass_Prelude.Apply_0"]())["__superclass_Prelude.Functor_0"]())(Node3.create(_522.value0))(f(_522.value1)))(f(_522.value2)))(f(_522.value3));
            };
            throw new Error("Failed pattern match");
        };
    };
});
var foldableFingerTree = new Data_Foldable.Foldable(function (__dict_Monoid_20) {
    return function (f) {
        return function (xs) {
            return Data_Foldable.foldr(foldableFingerTree)(function (x) {
                return function (acc) {
                    return Prelude["<>"](__dict_Monoid_20["__superclass_Prelude.Semigroup_0"]())(f(x))(acc);
                };
            })(Data_Monoid.mempty(__dict_Monoid_20))(xs);
        };
    };
}, function ($greater$minus) {
    return function (z) {
        return function (_527) {
            if (_527 instanceof Empty) {
                return z;
            };
            if (_527 instanceof Single) {
                return $greater$minus(z)(_527.value0);
            };
            if (_527 instanceof Deep) {
                var $greater$greater$greater$minus = Data_Foldable.foldl(foldableFingerTree)(Data_Foldable.foldl(foldableNode)($greater$minus));
                var $greater$greater$minus = Data_Foldable.foldl(Data_Foldable.foldableArray)($greater$minus);
                return $greater$greater$minus($greater$greater$greater$minus($greater$greater$minus(z)(_527.value1))(Data_Lazy.force(_527.value2)))(_527.value3);
            };
            throw new Error("Failed pattern match");
        };
    };
}, function ($minus$less) {
    return function (z) {
        return function (_526) {
            if (_526 instanceof Empty) {
                return z;
            };
            if (_526 instanceof Single) {
                return $minus$less(_526.value0)(z);
            };
            if (_526 instanceof Deep) {
                var $minus$less$less$less = Prelude.flip(Data_Foldable.foldr(foldableFingerTree)(Prelude.flip(Data_Foldable.foldr(foldableNode)($minus$less))));
                var $minus$less$less = Prelude.flip(Data_Foldable.foldr(Data_Foldable.foldableArray)($minus$less));
                var $plus$less$less = Prelude.flip(Data_Foldable.foldr(Data_Foldable.foldableArray)($minus$less));
                return $plus$less$less(_526.value1)($minus$less$less$less(Data_Lazy.force(_526.value2))($minus$less$less(_526.value3)(z)));
            };
            throw new Error("Failed pattern match");
        };
    };
});
var traversableFingerTree = new Data_Traversable.Traversable(function () {
    return foldableFingerTree;
}, function () {
    return functorFingerTree;
}, function (__dict_Applicative_3) {
    return Data_Traversable.traverse(traversableFingerTree)(__dict_Applicative_3)(Prelude.id(Prelude.categoryArr));
}, function (__dict_Applicative_2) {
    return function (f) {
        return function (_528) {
            if (_528 instanceof Empty) {
                return Prelude.pure(__dict_Applicative_2)(Empty.value);
            };
            if (_528 instanceof Single) {
                return Prelude["<$>"]((__dict_Applicative_2["__superclass_Prelude.Apply_0"]())["__superclass_Prelude.Functor_0"]())(Single.create)(f(_528.value0));
            };
            if (_528 instanceof Deep) {
                var l = Data_Traversable.traverse(traversableFingerTree)(__dict_Applicative_2)(Data_Traversable.traverse(traversableNode)(__dict_Applicative_2)(f))(Data_Lazy.force(_528.value2));
                var kl = Prelude["<$>"]((__dict_Applicative_2["__superclass_Prelude.Apply_0"]())["__superclass_Prelude.Functor_0"]())(Prelude["const"])(l);
                return Prelude["<*>"](__dict_Applicative_2["__superclass_Prelude.Apply_0"]())(Prelude["<*>"](__dict_Applicative_2["__superclass_Prelude.Apply_0"]())(Prelude["<$>"]((__dict_Applicative_2["__superclass_Prelude.Apply_0"]())["__superclass_Prelude.Functor_0"]())(Deep.create(_528.value0))(Data_Traversable.traverse(Data_Traversable.traversableArray)(__dict_Applicative_2)(f)(_528.value1)))(Prelude["<$>"]((__dict_Applicative_2["__superclass_Prelude.Apply_0"]())["__superclass_Prelude.Functor_0"]())(Data_Lazy.defer)(kl)))(Data_Traversable.traverse(Data_Traversable.traversableArray)(__dict_Applicative_2)(f)(_528.value3));
            };
            throw new Error("Failed pattern match");
        };
    };
});
var deep = function (__dict_Monoid_21) {
    return function (__dict_Measured_22) {
        return function (pr) {
            return function (m) {
                return function (sf) {
                    return new Deep(Data_Lazy.defer(function (_493) {
                        return Prelude["<>"](__dict_Monoid_21["__superclass_Prelude.Semigroup_0"]())(Data_Sequence_Internal.measure(Data_Sequence_Internal.measuredArray(__dict_Monoid_21)(__dict_Measured_22))(pr))(Prelude["<>"](__dict_Monoid_21["__superclass_Prelude.Semigroup_0"]())(Data_Sequence_Internal.measure(Data_Sequence_Internal.measuredLazy(__dict_Monoid_21)(measuredFingerTree(__dict_Monoid_21)(measuredNode)))(m))(Data_Sequence_Internal.measure(Data_Sequence_Internal.measuredArray(__dict_Monoid_21)(__dict_Measured_22))(sf)));
                    }), pr, m, sf);
                };
            };
        };
    };
};
var snoc = function (__dict_Monoid_23) {
    return function (__dict_Measured_24) {
        return function (_508) {
            return function (a) {
                if (_508 instanceof Empty) {
                    return new Single(a);
                };
                if (_508 instanceof Single) {
                    return deep(__dict_Monoid_23)(__dict_Measured_24)([ _508.value0 ])(lazyEmpty)([ a ]);
                };
                if (_508 instanceof Deep && _508.value3.length === 4) {
                    var forcedM = Data_Lazy.force(_508.value2);
                    return deep(__dict_Monoid_23)(__dict_Measured_24)(_508.value1)(Data_Lazy.defer(function (_495) {
                        return snoc(__dict_Monoid_23)(measuredNode)(forcedM)(node3(__dict_Monoid_23)(__dict_Measured_24)(_508.value3[0])(_508.value3[1])(_508.value3[2]));
                    }))([ _508.value3[3], a ]);
                };
                if (_508 instanceof Deep) {
                    return deep(__dict_Monoid_23)(__dict_Measured_24)(_508.value1)(_508.value2)(Data_Array.snoc(_508.value3)(a));
                };
                throw new Error("Failed pattern match");
            };
        };
    };
};
var snocAll = function (__dict_Monoid_25) {
    return function (__dict_Measured_26) {
        return function (__dict_Foldable_27) {
            return Data_Foldable.foldl(__dict_Foldable_27)(snoc(__dict_Monoid_25)(__dict_Measured_26));
        };
    };
};
var toFingerTree = function (__dict_Monoid_28) {
    return function (__dict_Measured_29) {
        return function (__dict_Foldable_30) {
            return function (s) {
                return snocAll(__dict_Monoid_28)(__dict_Measured_29)(__dict_Foldable_30)(Empty.value)(s);
            };
        };
    };
};
var viewL = function (__dict_Monoid_31) {
    return function (__dict_Measured_32) {
        return function (_509) {
            if (_509 instanceof Empty) {
                return NilL.value;
            };
            if (_509 instanceof Single) {
                return new ConsL(_509.value0, lazyEmpty);
            };
            if (_509 instanceof Deep) {
                return new ConsL(headDigit(_509.value1), Data_Lazy.defer(function (_496) {
                    return deepL(__dict_Monoid_31)(__dict_Measured_32)(tailDigit(_509.value1))(_509.value2)(_509.value3);
                }));
            };
            throw new Error("Failed pattern match");
        };
    };
};
var deepL = function (__dict_Monoid_33) {
    return function (__dict_Measured_34) {
        return function (_510) {
            return function (m) {
                return function (sf) {
                    if (_510.length === 0) {
                        var _2492 = viewL(__dict_Monoid_33)(measuredNode)(Data_Lazy.force(m));
                        if (_2492 instanceof NilL) {
                            return toFingerTree(__dict_Monoid_33)(__dict_Measured_34)(Data_Foldable.foldableArray)(sf);
                        };
                        if (_2492 instanceof ConsL) {
                            return deep(__dict_Monoid_33)(__dict_Measured_34)(nodeToDigit(_2492.value0))(_2492.value1)(sf);
                        };
                        throw new Error("Failed pattern match");
                    };
                    return deep(__dict_Monoid_33)(__dict_Measured_34)(_510)(m)(sf);
                };
            };
        };
    };
};
var eqFingerTree = function (__copy___dict_Monoid_35) {
    return function (__copy___dict_Measured_36) {
        return function (__copy___dict_Eq_37) {
            return function (__copy_xs) {
                return function (__copy_ys) {
                    var __dict_Monoid_35 = __copy___dict_Monoid_35;
                    var __dict_Measured_36 = __copy___dict_Measured_36;
                    var __dict_Eq_37 = __copy___dict_Eq_37;
                    var xs = __copy_xs;
                    var ys = __copy_ys;
                    tco: while (true) {
                        var _2495 = new Data_Tuple.Tuple(viewL(__dict_Monoid_35)(__dict_Measured_36)(xs), viewL(__dict_Monoid_35)(__dict_Measured_36)(ys));
                        if (_2495.value0 instanceof NilL && _2495.value1 instanceof NilL) {
                            return true;
                        };
                        if (_2495.value0 instanceof NilL) {
                            return false;
                        };
                        if (_2495.value1 instanceof NilL) {
                            return false;
                        };
                        if (_2495.value0 instanceof ConsL && _2495.value1 instanceof ConsL) {
                            var _2502 = Prelude["=="](__dict_Eq_37)(_2495.value0.value0)(_2495.value1.value0);
                            if (_2502) {
                                var ys$prime$prime = Data_Lazy.force(_2495.value1.value1);
                                var xs$prime$prime = Data_Lazy.force(_2495.value0.value1);
                                var __tco___dict_Monoid_35 = __dict_Monoid_35;
                                var __tco___dict_Measured_36 = __dict_Measured_36;
                                var __tco___dict_Eq_37 = __dict_Eq_37;
                                __dict_Monoid_35 = __tco___dict_Monoid_35;
                                __dict_Measured_36 = __tco___dict_Measured_36;
                                __dict_Eq_37 = __tco___dict_Eq_37;
                                xs = xs$prime$prime;
                                ys = ys$prime$prime;
                                continue tco;
                            };
                            if (!_2502) {
                                return false;
                            };
                            throw new Error("Failed pattern match");
                        };
                        throw new Error("Failed pattern match");
                    };
                };
            };
        };
    };
};
var head = function (__dict_Monoid_38) {
    return function (__dict_Measured_39) {
        return function (x) {
            var _2509 = viewL(__dict_Monoid_38)(__dict_Measured_39)(x);
            if (_2509 instanceof ConsL) {
                return new Data_Maybe.Just(_2509.value0);
            };
            if (_2509 instanceof NilL) {
                return Data_Maybe.Nothing.value;
            };
            throw new Error("Failed pattern match");
        };
    };
};
var isEmpty = function (__dict_Monoid_40) {
    return function (__dict_Measured_41) {
        return function (x) {
            var _2512 = viewL(__dict_Monoid_40)(__dict_Measured_41)(x);
            if (_2512 instanceof NilL) {
                return true;
            };
            if (_2512 instanceof ConsL) {
                return false;
            };
            throw new Error("Failed pattern match");
        };
    };
};
var tail = function (__dict_Monoid_42) {
    return function (__dict_Measured_43) {
        return function (x) {
            var _2515 = viewL(__dict_Monoid_42)(__dict_Measured_43)(x);
            if (_2515 instanceof ConsL) {
                return new Data_Maybe.Just(Data_Lazy.force(_2515.value1));
            };
            if (_2515 instanceof NilL) {
                return Data_Maybe.Nothing.value;
            };
            throw new Error("Failed pattern match");
        };
    };
};
var unfoldLeft = function (__dict_Unfoldable_44) {
    return function (__dict_Monoid_45) {
        return function (__dict_Measured_46) {
            var step = function (tree) {
                var _2518 = viewL(__dict_Monoid_45)(__dict_Measured_46)(tree);
                if (_2518 instanceof ConsL) {
                    return new Data_Maybe.Just(new Data_Tuple.Tuple(_2518.value0, Data_Lazy.force(_2518.value1)));
                };
                if (_2518 instanceof NilL) {
                    return Data_Maybe.Nothing.value;
                };
                throw new Error("Failed pattern match");
            };
            return Data_Unfoldable.unfoldr(__dict_Unfoldable_44)(step);
        };
    };
};
var viewR = function (__dict_Monoid_47) {
    return function (__dict_Measured_48) {
        return function (_511) {
            if (_511 instanceof Empty) {
                return NilR.value;
            };
            if (_511 instanceof Single) {
                return new SnocR(lazyEmpty, _511.value0);
            };
            if (_511 instanceof Deep) {
                return new SnocR(Data_Lazy.defer(function (_497) {
                    return deepR(__dict_Monoid_47)(__dict_Measured_48)(_511.value1)(_511.value2)(initDigit(_511.value3));
                }), lastDigit(_511.value3));
            };
            throw new Error("Failed pattern match");
        };
    };
};
var deepR = function (__dict_Monoid_49) {
    return function (__dict_Measured_50) {
        return function (pr) {
            return function (m) {
                return function (_512) {
                    if (_512.length === 0) {
                        var _2531 = viewR(__dict_Monoid_49)(measuredNode)(Data_Lazy.force(m));
                        if (_2531 instanceof NilR) {
                            return toFingerTree(__dict_Monoid_49)(__dict_Measured_50)(Data_Foldable.foldableArray)(pr);
                        };
                        if (_2531 instanceof SnocR) {
                            return deep(__dict_Monoid_49)(__dict_Measured_50)(pr)(_2531.value0)(nodeToDigit(_2531.value1));
                        };
                        throw new Error("Failed pattern match");
                    };
                    return deep(__dict_Monoid_49)(__dict_Measured_50)(pr)(m)(_512);
                };
            };
        };
    };
};
var init = function (__dict_Monoid_51) {
    return function (__dict_Measured_52) {
        return function (x) {
            var _2534 = viewR(__dict_Monoid_51)(__dict_Measured_52)(x);
            if (_2534 instanceof SnocR) {
                return new Data_Maybe.Just(Data_Lazy.force(_2534.value0));
            };
            if (_2534 instanceof NilR) {
                return Data_Maybe.Nothing.value;
            };
            throw new Error("Failed pattern match");
        };
    };
};
var last = function (__dict_Monoid_53) {
    return function (__dict_Measured_54) {
        return function (x) {
            var _2537 = viewR(__dict_Monoid_53)(__dict_Measured_54)(x);
            if (_2537 instanceof SnocR) {
                return new Data_Maybe.Just(_2537.value1);
            };
            if (_2537 instanceof NilR) {
                return Data_Maybe.Nothing.value;
            };
            throw new Error("Failed pattern match");
        };
    };
};
var unfoldRight = function (__dict_Unfoldable_55) {
    return function (__dict_Monoid_56) {
        return function (__dict_Measured_57) {
            var step = function (tree) {
                var _2540 = viewR(__dict_Monoid_56)(__dict_Measured_57)(tree);
                if (_2540 instanceof SnocR) {
                    return new Data_Maybe.Just(new Data_Tuple.Tuple(_2540.value1, Data_Lazy.force(_2540.value0)));
                };
                if (_2540 instanceof NilR) {
                    return Data_Maybe.Nothing.value;
                };
                throw new Error("Failed pattern match");
            };
            return Data_Unfoldable.unfoldr(__dict_Unfoldable_55)(step);
        };
    };
};
var unsafeSplitTree = function (__dict_Monoid_58) {
    return function (__dict_Measured_59) {
        return function (p) {
            return function (i) {
                return function (_516) {
                    if (_516 instanceof Single) {
                        return new LazySplit(lazyEmpty, _516.value0, lazyEmpty);
                    };
                    if (_516 instanceof Deep) {
                        var vpr = Prelude["<>"](__dict_Monoid_58["__superclass_Prelude.Semigroup_0"]())(i)(Data_Sequence_Internal.measure(Data_Sequence_Internal.measuredArray(__dict_Monoid_58)(__dict_Measured_59))(_516.value1));
                        var _2547 = p(vpr);
                        if (_2547) {
                            var _2548 = splitDigit(__dict_Monoid_58)(__dict_Measured_59)(p)(i)(_516.value1);
                            return new LazySplit(Data_Lazy.defer(function (_498) {
                                return toFingerTree(__dict_Monoid_58)(__dict_Measured_59)(Data_Foldable.foldableArray)(_2548.value0);
                            }), _2548.value1, Data_Lazy.defer(function (_499) {
                                return deepL(__dict_Monoid_58)(__dict_Measured_59)(_2548.value2)(_516.value2)(_516.value3);
                            }));
                        };
                        if (!_2547) {
                            var vm = Prelude["<>"](__dict_Monoid_58["__superclass_Prelude.Semigroup_0"]())(vpr)(Data_Sequence_Internal.measure(Data_Sequence_Internal.measuredLazy(__dict_Monoid_58)(measuredFingerTree(__dict_Monoid_58)(measuredNode)))(_516.value2));
                            var _2554 = p(vm);
                            if (_2554) {
                                var _2555 = unsafeSplitTree(__dict_Monoid_58)(measuredNode)(p)(vpr)(Data_Lazy.force(_516.value2));
                                var _2556 = splitDigit(__dict_Monoid_58)(__dict_Measured_59)(p)(Prelude["<>"](__dict_Monoid_58["__superclass_Prelude.Semigroup_0"]())(vpr)(Data_Sequence_Internal.measure(Data_Sequence_Internal.measuredLazy(__dict_Monoid_58)(measuredFingerTree(__dict_Monoid_58)(measuredNode)))(_2555.value0)))(nodeToDigit(_2555.value1));
                                return new LazySplit(Data_Lazy.defer(function (_500) {
                                    return deepR(__dict_Monoid_58)(__dict_Measured_59)(_516.value1)(_2555.value0)(_2556.value0);
                                }), _2556.value1, Data_Lazy.defer(function (_501) {
                                    return deepL(__dict_Monoid_58)(__dict_Measured_59)(_2556.value2)(_2555.value2)(_516.value3);
                                }));
                            };
                            if (!_2554) {
                                var _2565 = splitDigit(__dict_Monoid_58)(__dict_Measured_59)(p)(vm)(_516.value3);
                                return new LazySplit(Data_Lazy.defer(function (_502) {
                                    return deepR(__dict_Monoid_58)(__dict_Measured_59)(_516.value1)(_516.value2)(_2565.value0);
                                }), _2565.value1, Data_Lazy.defer(function (_503) {
                                    return toFingerTree(__dict_Monoid_58)(__dict_Measured_59)(Data_Foldable.foldableArray)(_2565.value2);
                                }));
                            };
                            throw new Error("Failed pattern match");
                        };
                        throw new Error("Failed pattern match");
                    };
                    throw new Error("Failed pattern match");
                };
            };
        };
    };
};
var cons = function (__dict_Monoid_60) {
    return function (__dict_Measured_61) {
        return function (a) {
            return function (_507) {
                if (_507 instanceof Empty) {
                    return new Single(a);
                };
                if (_507 instanceof Single) {
                    return deep(__dict_Monoid_60)(__dict_Measured_61)([ a ])(lazyEmpty)([ _507.value0 ]);
                };
                if (_507 instanceof Deep && _507.value1.length === 4) {
                    var forcedM = Data_Lazy.force(_507.value2);
                    return deep(__dict_Monoid_60)(__dict_Measured_61)([ a, _507.value1[0] ])(Data_Lazy.defer(function (_494) {
                        return cons(__dict_Monoid_60)(measuredNode)(node3(__dict_Monoid_60)(__dict_Measured_61)(_507.value1[1])(_507.value1[2])(_507.value1[3]))(forcedM);
                    }))(_507.value3);
                };
                if (_507 instanceof Deep) {
                    return deep(__dict_Monoid_60)(__dict_Measured_61)(Prelude[":"](a)(_507.value1))(_507.value2)(_507.value3);
                };
                throw new Error("Failed pattern match");
            };
        };
    };
};
var consAll = function (__dict_Monoid_62) {
    return function (__dict_Measured_63) {
        return function (__dict_Foldable_64) {
            return Prelude.flip(Data_Foldable.foldr(__dict_Foldable_64)(cons(__dict_Monoid_62)(__dict_Measured_63)));
        };
    };
};
var filter = function (__dict_Monoid_65) {
    return function (__dict_Measured_66) {
        return function (p) {
            return Data_Foldable.foldr(foldableFingerTree)(function (x) {
                return function (acc) {
                    var _2591 = p(x);
                    if (_2591) {
                        return cons(__dict_Monoid_65)(__dict_Measured_66)(x)(acc);
                    };
                    if (!_2591) {
                        return acc;
                    };
                    throw new Error("Failed pattern match");
                };
            })(Empty.value);
        };
    };
};
var split = function (__dict_Monoid_67) {
    return function (__dict_Measured_68) {
        return function (p) {
            return function (_517) {
                if (_517 instanceof Empty) {
                    return new Data_Tuple.Tuple(lazyEmpty, lazyEmpty);
                };
                var _2594 = p(Data_Sequence_Internal.measure(measuredFingerTree(__dict_Monoid_67)(__dict_Measured_68))(_517));
                if (_2594) {
                    var _2595 = unsafeSplitTree(__dict_Monoid_67)(__dict_Measured_68)(p)(Data_Monoid.mempty(__dict_Monoid_67))(_517);
                    return new Data_Tuple.Tuple(_2595.value0, Data_Lazy.defer(function (_504) {
                        return cons(__dict_Monoid_67)(__dict_Measured_68)(_2595.value1)(Data_Lazy.force(_2595.value2));
                    }));
                };
                if (!_2594) {
                    return new Data_Tuple.Tuple(Data_Lazy.defer(function (_505) {
                        return _517;
                    }), lazyEmpty);
                };
                throw new Error("Failed pattern match");
            };
        };
    };
};
var compareFingerTree = function (__copy___dict_Monoid_69) {
    return function (__copy___dict_Measured_70) {
        return function (__copy___dict_Ord_71) {
            return function (__copy_xs) {
                return function (__copy_ys) {
                    var __dict_Monoid_69 = __copy___dict_Monoid_69;
                    var __dict_Measured_70 = __copy___dict_Measured_70;
                    var __dict_Ord_71 = __copy___dict_Ord_71;
                    var xs = __copy_xs;
                    var ys = __copy_ys;
                    tco: while (true) {
                        var _2601 = new Data_Tuple.Tuple(viewL(__dict_Monoid_69)(__dict_Measured_70)(xs), viewL(__dict_Monoid_69)(__dict_Measured_70)(ys));
                        if (_2601.value0 instanceof NilL && _2601.value1 instanceof NilL) {
                            return Prelude.EQ.value;
                        };
                        if (_2601.value0 instanceof NilL) {
                            return Prelude.LT.value;
                        };
                        if (_2601.value1 instanceof NilL) {
                            return Prelude.GT.value;
                        };
                        if (_2601.value0 instanceof ConsL && _2601.value1 instanceof ConsL) {
                            var _2608 = Prelude.compare(__dict_Ord_71)(_2601.value0.value0)(_2601.value1.value0);
                            if (_2608 instanceof Prelude.EQ) {
                                var ys$prime$prime = Data_Lazy.force(_2601.value1.value1);
                                var xs$prime$prime = Data_Lazy.force(_2601.value0.value1);
                                var __tco___dict_Monoid_69 = __dict_Monoid_69;
                                var __tco___dict_Measured_70 = __dict_Measured_70;
                                var __tco___dict_Ord_71 = __dict_Ord_71;
                                __dict_Monoid_69 = __tco___dict_Monoid_69;
                                __dict_Measured_70 = __tco___dict_Measured_70;
                                __dict_Ord_71 = __tco___dict_Ord_71;
                                xs = xs$prime$prime;
                                ys = ys$prime$prime;
                                continue tco;
                            };
                            return _2608;
                        };
                        throw new Error("Failed pattern match");
                    };
                };
            };
        };
    };
};
var app3 = function (__dict_Monoid_72) {
    return function (__dict_Measured_73) {
        return function (_513) {
            return function (ts) {
                return function (_514) {
                    if (_513 instanceof Empty) {
                        return consAll(__dict_Monoid_72)(__dict_Measured_73)(Data_Foldable.foldableArray)(ts)(_514);
                    };
                    if (_514 instanceof Empty) {
                        return snocAll(__dict_Monoid_72)(__dict_Measured_73)(Data_Foldable.foldableArray)(_513)(ts);
                    };
                    if (_513 instanceof Single) {
                        return cons(__dict_Monoid_72)(__dict_Measured_73)(_513.value0)(consAll(__dict_Monoid_72)(__dict_Measured_73)(Data_Foldable.foldableArray)(ts)(_514));
                    };
                    if (_514 instanceof Single) {
                        return snoc(__dict_Monoid_72)(__dict_Measured_73)(snocAll(__dict_Monoid_72)(__dict_Measured_73)(Data_Foldable.foldableArray)(_513)(ts))(_514.value0);
                    };
                    if (_513 instanceof Deep && _514 instanceof Deep) {
                        var computeM$prime = function (_531) {
                            return app3(__dict_Monoid_72)(measuredNode)(Data_Lazy.force(_513.value2))(nodes(__dict_Monoid_72)(__dict_Measured_73)(Prelude["<>"](Data_Array.semigroupArray)(_513.value3)(Prelude["<>"](Data_Array.semigroupArray)(ts)(_514.value1))))(Data_Lazy.force(_514.value2));
                        };
                        return deep(__dict_Monoid_72)(__dict_Measured_73)(_513.value1)(Data_Lazy.defer(computeM$prime))(_514.value3);
                    };
                    throw new Error("Failed pattern match");
                };
            };
        };
    };
};
var append = function (__dict_Monoid_74) {
    return function (__dict_Measured_75) {
        return function (xs) {
            return function (ys) {
                return app3(__dict_Monoid_74)(__dict_Measured_75)(xs)([  ])(ys);
            };
        };
    };
};
var semigroupFingerTree = function (__dict_Monoid_76) {
    return function (__dict_Measured_77) {
        return new Prelude.Semigroup(append(__dict_Monoid_76)(__dict_Measured_77));
    };
};
module.exports = {
    LazySplit: LazySplit, 
    Split: Split, 
    NilR: NilR, 
    SnocR: SnocR, 
    NilL: NilL, 
    ConsL: ConsL, 
    Empty: Empty, 
    Single: Single, 
    Deep: Deep, 
    Node2: Node2, 
    Node3: Node3, 
    unfoldRight: unfoldRight, 
    unfoldLeft: unfoldLeft, 
    filter: filter, 
    split: split, 
    unsafeSplitTree: unsafeSplitTree, 
    splitDigit: splitDigit, 
    append: append, 
    nodes: nodes, 
    app3: app3, 
    init: init, 
    last: last, 
    deepR: deepR, 
    viewR: viewR, 
    initDigit: initDigit, 
    lastDigit: lastDigit, 
    tail: tail, 
    head: head, 
    isEmpty: isEmpty, 
    deepL: deepL, 
    viewL: viewL, 
    tailDigit: tailDigit, 
    headDigit: headDigit, 
    toFingerTree: toFingerTree, 
    snocAll: snocAll, 
    consAll: consAll, 
    snoc: snoc, 
    cons: cons, 
    compareFingerTree: compareFingerTree, 
    eqFingerTree: eqFingerTree, 
    deep: deep, 
    lazyEmpty: lazyEmpty, 
    nodeToDigit: nodeToDigit, 
    node3: node3, 
    node2: node2, 
    showNode: showNode, 
    functorNode: functorNode, 
    foldableNode: foldableNode, 
    traversableNode: traversableNode, 
    measuredNode: measuredNode, 
    showFingerTree: showFingerTree, 
    semigroupFingerTree: semigroupFingerTree, 
    functorFingerTree: functorFingerTree, 
    foldableFingerTree: foldableFingerTree, 
    traversableFingerTree: traversableFingerTree, 
    measuredFingerTree: measuredFingerTree, 
    functorViewL: functorViewL
};
