// Data.FingerTree/index.js
// Generated by psc-make version 0.6.9.5
"use strict";
var Prelude = require("Prelude");
var Data_Sequence_Internal = require("Data.Sequence.Internal");
var Data_Foldable = require("Data.Foldable");
var Data_Monoid = require("Data.Monoid");
var Data_Traversable = require("Data.Traversable");
var Data_Lazy = require("Data.Lazy");
var Data_Array = require("Data.Array");
var Data_Array_Unsafe = require("Data.Array.Unsafe");
var Data_Unfoldable = require("Data.Unfoldable");
var Data_Maybe = require("Data.Maybe");
var Data_Tuple = require("Data.Tuple");
var NilR = (function () {
    function NilR() {

    };
    NilR.value = new NilR();
    return NilR;
})();
var SnocR = (function () {
    function SnocR(value0, value1) {
        this.value0 = value0;
        this.value1 = value1;
    };
    SnocR.create = function (value0) {
        return function (value1) {
            return new SnocR(value0, value1);
        };
    };
    return SnocR;
})();
var NilL = (function () {
    function NilL() {

    };
    NilL.value = new NilL();
    return NilL;
})();
var ConsL = (function () {
    function ConsL(value0, value1) {
        this.value0 = value0;
        this.value1 = value1;
    };
    ConsL.create = function (value0) {
        return function (value1) {
            return new ConsL(value0, value1);
        };
    };
    return ConsL;
})();
var Split = (function () {
    function Split(value0, value1, value2) {
        this.value0 = value0;
        this.value1 = value1;
        this.value2 = value2;
    };
    Split.create = function (value0) {
        return function (value1) {
            return function (value2) {
                return new Split(value0, value1, value2);
            };
        };
    };
    return Split;
})();
var Node2 = (function () {
    function Node2(value0, value1, value2) {
        this.value0 = value0;
        this.value1 = value1;
        this.value2 = value2;
    };
    Node2.create = function (value0) {
        return function (value1) {
            return function (value2) {
                return new Node2(value0, value1, value2);
            };
        };
    };
    return Node2;
})();
var Node3 = (function () {
    function Node3(value0, value1, value2, value3) {
        this.value0 = value0;
        this.value1 = value1;
        this.value2 = value2;
        this.value3 = value3;
    };
    Node3.create = function (value0) {
        return function (value1) {
            return function (value2) {
                return function (value3) {
                    return new Node3(value0, value1, value2, value3);
                };
            };
        };
    };
    return Node3;
})();
var LazySplit = (function () {
    function LazySplit(value0, value1, value2) {
        this.value0 = value0;
        this.value1 = value1;
        this.value2 = value2;
    };
    LazySplit.create = function (value0) {
        return function (value1) {
            return function (value2) {
                return new LazySplit(value0, value1, value2);
            };
        };
    };
    return LazySplit;
})();
var Empty = (function () {
    function Empty() {

    };
    Empty.value = new Empty();
    return Empty;
})();
var Single = (function () {
    function Single(value0) {
        this.value0 = value0;
    };
    Single.create = function (value0) {
        return new Single(value0);
    };
    return Single;
})();
var Deep = (function () {
    function Deep(value0, value1, value2, value3) {
        this.value0 = value0;
        this.value1 = value1;
        this.value2 = value2;
        this.value3 = value3;
    };
    Deep.create = function (value0) {
        return function (value1) {
            return function (value2) {
                return function (value3) {
                    return new Deep(value0, value1, value2, value3);
                };
            };
        };
    };
    return Deep;
})();
var tailDigit = Data_Array_Unsafe.tail;
var splitDigit = function (__dict_Monoid_4) {
    return function (__dict_Measured_5) {
        return function (p) {
            return function (i) {
                return function (as) {
                    var _2389 = Data_Array.length(as);
                    if (_2389 === 1) {
                        return new Split([  ], Data_Sequence_Internal["!"](as)(0), [  ]);
                    };
                    var bs = Data_Array.drop(1)(as);
                    var a = Data_Sequence_Internal["!"](as)(0);
                    var i$prime = Prelude["<>"](__dict_Monoid_4["__superclass_Prelude.Semigroup_0"]())(i)(Data_Sequence_Internal.measure(__dict_Measured_5)(a));
                    var _2390 = p(i$prime);
                    if (_2390) {
                        return new Split([  ], a, bs);
                    };
                    if (!_2390) {
                        var _2391 = splitDigit(__dict_Monoid_4)(__dict_Measured_5)(p)(i$prime)(bs);
                        return new Split(Prelude[":"](a)(_2391.value0), _2391.value1, _2391.value2);
                    };
                    throw new Error("Failed pattern match");
                };
            };
        };
    };
};
var showNode = function (__dict_Show_6) {
    return function (__dict_Show_7) {
        return new Prelude.Show(function (_529) {
            if (_529 instanceof Node2) {
                return "Node2 (" + (Prelude.show(__dict_Show_7)(_529.value0) + (") (" + (Prelude.show(__dict_Show_6)(_529.value1) + (") (" + (Prelude.show(__dict_Show_6)(_529.value2) + ")")))));
            };
            if (_529 instanceof Node3) {
                return "Node3 (" + (Prelude.show(__dict_Show_7)(_529.value0) + (") (" + (Prelude.show(__dict_Show_6)(_529.value1) + (") (" + (Prelude.show(__dict_Show_6)(_529.value2) + (") (" + (Prelude.show(__dict_Show_6)(_529.value3) + ")")))))));
            };
            throw new Error("Failed pattern match");
        });
    };
};
var showFingerTree = function (__dict_Show_8) {
    return function (__dict_Show_9) {
        return new Prelude.Show(function (_535) {
            if (_535 instanceof Empty) {
                return "Empty";
            };
            if (_535 instanceof Single) {
                return "Single (" + (Prelude.show(__dict_Show_9)(_535.value0) + ")");
            };
            if (_535 instanceof Deep) {
                return "Deep (" + (Prelude.show(Data_Lazy.showLazy(__dict_Show_8))(_535.value0) + (") (" + (Prelude.show(Prelude.showArray(__dict_Show_9))(_535.value1) + (") (" + (Prelude.show(Data_Lazy.showLazy(showFingerTree(__dict_Show_8)(showNode(__dict_Show_9)(__dict_Show_8))))(_535.value2) + (") (" + (Prelude.show(Prelude.showArray(__dict_Show_9))(_535.value3) + ")")))))));
            };
            throw new Error("Failed pattern match");
        });
    };
};
var nodeToDigit = function (_517) {
    if (_517 instanceof Node2) {
        return [ _517.value1, _517.value2 ];
    };
    if (_517 instanceof Node3) {
        return [ _517.value1, _517.value2, _517.value3 ];
    };
    throw new Error("Failed pattern match");
};
var node3 = function (__dict_Monoid_10) {
    return function (__dict_Measured_11) {
        return function (a) {
            return function (b) {
                return function (c) {
                    return new Node3(Prelude["<>"](__dict_Monoid_10["__superclass_Prelude.Semigroup_0"]())(Data_Sequence_Internal.measure(__dict_Measured_11)(a))(Prelude["<>"](__dict_Monoid_10["__superclass_Prelude.Semigroup_0"]())(Data_Sequence_Internal.measure(__dict_Measured_11)(b))(Data_Sequence_Internal.measure(__dict_Measured_11)(c))), a, b, c);
                };
            };
        };
    };
};
var node2 = function (__dict_Monoid_12) {
    return function (__dict_Measured_13) {
        return function (a) {
            return function (b) {
                return new Node2(Prelude["<>"](__dict_Monoid_12["__superclass_Prelude.Semigroup_0"]())(Data_Sequence_Internal.measure(__dict_Measured_13)(a))(Data_Sequence_Internal.measure(__dict_Measured_13)(b)), a, b);
            };
        };
    };
};
var nodes = function (__dict_Monoid_14) {
    return function (__dict_Measured_15) {
        return function (_526) {
            if (_526.length === 2) {
                return [ node2(__dict_Monoid_14)(__dict_Measured_15)(_526[0])(_526[1]) ];
            };
            if (_526.length === 3) {
                return [ node3(__dict_Monoid_14)(__dict_Measured_15)(_526[0])(_526[1])(_526[2]) ];
            };
            if (_526.length === 4) {
                return [ node2(__dict_Monoid_14)(__dict_Measured_15)(_526[0])(_526[1]), node2(__dict_Monoid_14)(__dict_Measured_15)(_526[2])(_526[3]) ];
            };
            return Prelude[":"](node3(__dict_Monoid_14)(__dict_Measured_15)(Data_Sequence_Internal["!"](_526)(0))(Data_Sequence_Internal["!"](_526)(1))(Data_Sequence_Internal["!"](_526)(2)))(nodes(__dict_Monoid_14)(__dict_Measured_15)(Data_Array.drop(3)(_526)));
        };
    };
};
var measuredNode = new Data_Sequence_Internal.Measured(function (_534) {
    if (_534 instanceof Node2) {
        return _534.value0;
    };
    if (_534 instanceof Node3) {
        return _534.value0;
    };
    throw new Error("Failed pattern match");
});
var measuredFingerTree = function (__dict_Monoid_16) {
    return function (__dict_Measured_17) {
        return new Data_Sequence_Internal.Measured(function (_540) {
            if (_540 instanceof Empty) {
                return Data_Monoid.mempty(__dict_Monoid_16);
            };
            if (_540 instanceof Single) {
                return Data_Sequence_Internal.measure(__dict_Measured_17)(_540.value0);
            };
            if (_540 instanceof Deep) {
                return Data_Lazy.force(_540.value0);
            };
            throw new Error("Failed pattern match");
        });
    };
};
var lazyEmpty = Data_Lazy.defer(function (_503) {
    return Empty.value;
});
var lastDigit = Data_Array_Unsafe.last;
var initDigit = Data_Array_Unsafe.init;
var headDigit = Data_Array_Unsafe.head;
var functorViewL = function (__dict_Functor_18) {
    return new Prelude.Functor(function (f) {
        return function (_541) {
            if (_541 instanceof NilL) {
                return NilL.value;
            };
            if (_541 instanceof ConsL) {
                return new ConsL(f(_541.value0), Prelude["<$>"](Data_Lazy.functorLazy)(Prelude["<$>"](__dict_Functor_18)(f))(_541.value1));
            };
            throw new Error("Failed pattern match");
        };
    });
};
var functorNode = new Prelude.Functor(function (f) {
    return function (_530) {
        if (_530 instanceof Node2) {
            return new Node2(_530.value0, f(_530.value1), f(_530.value2));
        };
        if (_530 instanceof Node3) {
            return new Node3(_530.value0, f(_530.value1), f(_530.value2), f(_530.value3));
        };
        throw new Error("Failed pattern match");
    };
});
var functorFingerTree = new Prelude.Functor(function (f) {
    return function (_536) {
        if (_536 instanceof Empty) {
            return Empty.value;
        };
        if (_536 instanceof Single) {
            return new Single(f(_536.value0));
        };
        if (_536 instanceof Deep) {
            return new Deep(_536.value0, Prelude["<$>"](Data_Array.functorArray)(f)(_536.value1), Data_Sequence_Internal["<$$$>"](Data_Lazy.functorLazy)(functorFingerTree)(functorNode)(f)(_536.value2), Prelude["<$>"](Data_Array.functorArray)(f)(_536.value3));
        };
        throw new Error("Failed pattern match");
    };
});
var foldableNode = new Data_Foldable.Foldable(function (__dict_Monoid_19) {
    return function (f) {
        return function (xs) {
            return Data_Foldable.foldr(foldableNode)(function (x) {
                return function (acc) {
                    return Prelude["<>"](__dict_Monoid_19["__superclass_Prelude.Semigroup_0"]())(f(x))(acc);
                };
            })(Data_Monoid.mempty(__dict_Monoid_19))(xs);
        };
    };
}, function ($greater$minus) {
    return function (z) {
        return function (_532) {
            if (_532 instanceof Node2) {
                return $greater$minus($greater$minus(z)(_532.value1))(_532.value2);
            };
            if (_532 instanceof Node3) {
                return $greater$minus($greater$minus($greater$minus(z)(_532.value1))(_532.value2))(_532.value3);
            };
            throw new Error("Failed pattern match");
        };
    };
}, function ($minus$less) {
    return function (z) {
        return function (_531) {
            if (_531 instanceof Node2) {
                return $minus$less(_531.value1)($minus$less(_531.value2)(z));
            };
            if (_531 instanceof Node3) {
                return $minus$less(_531.value1)($minus$less(_531.value2)($minus$less(_531.value3)(z)));
            };
            throw new Error("Failed pattern match");
        };
    };
});
var traversableNode = new Data_Traversable.Traversable(function () {
    return foldableNode;
}, function () {
    return functorNode;
}, function (__dict_Applicative_1) {
    return Data_Traversable.traverse(traversableNode)(__dict_Applicative_1)(Prelude.id(Prelude.categoryArr));
}, function (__dict_Applicative_0) {
    return function (f) {
        return function (_533) {
            if (_533 instanceof Node2) {
                return Prelude["<*>"](__dict_Applicative_0["__superclass_Prelude.Apply_0"]())(Prelude["<$>"]((__dict_Applicative_0["__superclass_Prelude.Apply_0"]())["__superclass_Prelude.Functor_0"]())(Node2.create(_533.value0))(f(_533.value1)))(f(_533.value2));
            };
            if (_533 instanceof Node3) {
                return Prelude["<*>"](__dict_Applicative_0["__superclass_Prelude.Apply_0"]())(Prelude["<*>"](__dict_Applicative_0["__superclass_Prelude.Apply_0"]())(Prelude["<$>"]((__dict_Applicative_0["__superclass_Prelude.Apply_0"]())["__superclass_Prelude.Functor_0"]())(Node3.create(_533.value0))(f(_533.value1)))(f(_533.value2)))(f(_533.value3));
            };
            throw new Error("Failed pattern match");
        };
    };
});
var foldableFingerTree = new Data_Foldable.Foldable(function (__dict_Monoid_20) {
    return function (f) {
        return function (xs) {
            return Data_Foldable.foldr(foldableFingerTree)(function (x) {
                return function (acc) {
                    return Prelude["<>"](__dict_Monoid_20["__superclass_Prelude.Semigroup_0"]())(f(x))(acc);
                };
            })(Data_Monoid.mempty(__dict_Monoid_20))(xs);
        };
    };
}, function ($greater$minus) {
    return function (z) {
        return function (_538) {
            if (_538 instanceof Empty) {
                return z;
            };
            if (_538 instanceof Single) {
                return $greater$minus(z)(_538.value0);
            };
            if (_538 instanceof Deep) {
                var $greater$greater$greater$minus = Data_Foldable.foldl(foldableFingerTree)(Data_Foldable.foldl(foldableNode)($greater$minus));
                var $greater$greater$minus = Data_Foldable.foldl(Data_Foldable.foldableArray)($greater$minus);
                return $greater$greater$minus($greater$greater$greater$minus($greater$greater$minus(z)(_538.value1))(Data_Lazy.force(_538.value2)))(_538.value3);
            };
            throw new Error("Failed pattern match");
        };
    };
}, function ($minus$less) {
    return function (z) {
        return function (_537) {
            if (_537 instanceof Empty) {
                return z;
            };
            if (_537 instanceof Single) {
                return $minus$less(_537.value0)(z);
            };
            if (_537 instanceof Deep) {
                var $minus$less$less$less = Prelude.flip(Data_Foldable.foldr(foldableFingerTree)(Prelude.flip(Data_Foldable.foldr(foldableNode)($minus$less))));
                var $minus$less$less = Prelude.flip(Data_Foldable.foldr(Data_Foldable.foldableArray)($minus$less));
                var $plus$less$less = Prelude.flip(Data_Foldable.foldr(Data_Foldable.foldableArray)($minus$less));
                return $plus$less$less(_537.value1)($minus$less$less$less(Data_Lazy.force(_537.value2))($minus$less$less(_537.value3)(z)));
            };
            throw new Error("Failed pattern match");
        };
    };
});
var traversableFingerTree = new Data_Traversable.Traversable(function () {
    return foldableFingerTree;
}, function () {
    return functorFingerTree;
}, function (__dict_Applicative_3) {
    return Data_Traversable.traverse(traversableFingerTree)(__dict_Applicative_3)(Prelude.id(Prelude.categoryArr));
}, function (__dict_Applicative_2) {
    return function (f) {
        return function (_539) {
            if (_539 instanceof Empty) {
                return Prelude.pure(__dict_Applicative_2)(Empty.value);
            };
            if (_539 instanceof Single) {
                return Prelude["<$>"]((__dict_Applicative_2["__superclass_Prelude.Apply_0"]())["__superclass_Prelude.Functor_0"]())(Single.create)(f(_539.value0));
            };
            if (_539 instanceof Deep) {
                var l = Data_Traversable.traverse(traversableFingerTree)(__dict_Applicative_2)(Data_Traversable.traverse(traversableNode)(__dict_Applicative_2)(f))(Data_Lazy.force(_539.value2));
                var kl = Prelude["<$>"]((__dict_Applicative_2["__superclass_Prelude.Apply_0"]())["__superclass_Prelude.Functor_0"]())(Prelude["const"])(l);
                return Prelude["<*>"](__dict_Applicative_2["__superclass_Prelude.Apply_0"]())(Prelude["<*>"](__dict_Applicative_2["__superclass_Prelude.Apply_0"]())(Prelude["<$>"]((__dict_Applicative_2["__superclass_Prelude.Apply_0"]())["__superclass_Prelude.Functor_0"]())(Deep.create(_539.value0))(Data_Traversable.traverse(Data_Traversable.traversableArray)(__dict_Applicative_2)(f)(_539.value1)))(Prelude["<$>"]((__dict_Applicative_2["__superclass_Prelude.Apply_0"]())["__superclass_Prelude.Functor_0"]())(Data_Lazy.defer)(kl)))(Data_Traversable.traverse(Data_Traversable.traversableArray)(__dict_Applicative_2)(f)(_539.value3));
            };
            throw new Error("Failed pattern match");
        };
    };
});
var deep = function (__dict_Monoid_21) {
    return function (__dict_Measured_22) {
        return function (pr) {
            return function (m) {
                return function (sf) {
                    return new Deep(Data_Lazy.defer(function (_504) {
                        return Prelude["<>"](__dict_Monoid_21["__superclass_Prelude.Semigroup_0"]())(Data_Sequence_Internal.measure(Data_Sequence_Internal.measuredArray(__dict_Monoid_21)(__dict_Measured_22))(pr))(Prelude["<>"](__dict_Monoid_21["__superclass_Prelude.Semigroup_0"]())(Data_Sequence_Internal.measure(Data_Sequence_Internal.measuredLazy(__dict_Monoid_21)(measuredFingerTree(__dict_Monoid_21)(measuredNode)))(m))(Data_Sequence_Internal.measure(Data_Sequence_Internal.measuredArray(__dict_Monoid_21)(__dict_Measured_22))(sf)));
                    }), pr, m, sf);
                };
            };
        };
    };
};
var snoc = function (__dict_Monoid_23) {
    return function (__dict_Measured_24) {
        return function (_519) {
            return function (a) {
                if (_519 instanceof Empty) {
                    return new Single(a);
                };
                if (_519 instanceof Single) {
                    return deep(__dict_Monoid_23)(__dict_Measured_24)([ _519.value0 ])(lazyEmpty)([ a ]);
                };
                if (_519 instanceof Deep && _519.value3.length === 4) {
                    var forcedM = Data_Lazy.force(_519.value2);
                    return deep(__dict_Monoid_23)(__dict_Measured_24)(_519.value1)(Data_Lazy.defer(function (_506) {
                        return snoc(__dict_Monoid_23)(measuredNode)(forcedM)(node3(__dict_Monoid_23)(__dict_Measured_24)(_519.value3[0])(_519.value3[1])(_519.value3[2]));
                    }))([ _519.value3[3], a ]);
                };
                if (_519 instanceof Deep) {
                    return deep(__dict_Monoid_23)(__dict_Measured_24)(_519.value1)(_519.value2)(Data_Array.snoc(_519.value3)(a));
                };
                throw new Error("Failed pattern match");
            };
        };
    };
};
var snocAll = function (__dict_Monoid_25) {
    return function (__dict_Measured_26) {
        return function (__dict_Foldable_27) {
            return Data_Foldable.foldl(__dict_Foldable_27)(snoc(__dict_Monoid_25)(__dict_Measured_26));
        };
    };
};
var toFingerTree = function (__dict_Monoid_28) {
    return function (__dict_Measured_29) {
        return function (__dict_Foldable_30) {
            return function (s) {
                return snocAll(__dict_Monoid_28)(__dict_Measured_29)(__dict_Foldable_30)(Empty.value)(s);
            };
        };
    };
};
var viewL = function (__dict_Monoid_31) {
    return function (__dict_Measured_32) {
        return function (_520) {
            if (_520 instanceof Empty) {
                return NilL.value;
            };
            if (_520 instanceof Single) {
                return new ConsL(_520.value0, lazyEmpty);
            };
            if (_520 instanceof Deep) {
                return new ConsL(headDigit(_520.value1), Data_Lazy.defer(function (_507) {
                    return deepL(__dict_Monoid_31)(__dict_Measured_32)(tailDigit(_520.value1))(_520.value2)(_520.value3);
                }));
            };
            throw new Error("Failed pattern match");
        };
    };
};
var deepL = function (__dict_Monoid_33) {
    return function (__dict_Measured_34) {
        return function (_521) {
            return function (m) {
                return function (sf) {
                    if (_521.length === 0) {
                        var _2541 = viewL(__dict_Monoid_33)(measuredNode)(Data_Lazy.force(m));
                        if (_2541 instanceof NilL) {
                            return toFingerTree(__dict_Monoid_33)(__dict_Measured_34)(Data_Foldable.foldableArray)(sf);
                        };
                        if (_2541 instanceof ConsL) {
                            return deep(__dict_Monoid_33)(__dict_Measured_34)(nodeToDigit(_2541.value0))(_2541.value1)(sf);
                        };
                        throw new Error("Failed pattern match");
                    };
                    return deep(__dict_Monoid_33)(__dict_Measured_34)(_521)(m)(sf);
                };
            };
        };
    };
};
var eqFingerTree = function (__copy___dict_Monoid_35) {
    return function (__copy___dict_Measured_36) {
        return function (__copy___dict_Eq_37) {
            return function (__copy_xs) {
                return function (__copy_ys) {
                    var __dict_Monoid_35 = __copy___dict_Monoid_35;
                    var __dict_Measured_36 = __copy___dict_Measured_36;
                    var __dict_Eq_37 = __copy___dict_Eq_37;
                    var xs = __copy_xs;
                    var ys = __copy_ys;
                    tco: while (true) {
                        var _2544 = new Data_Tuple.Tuple(viewL(__dict_Monoid_35)(__dict_Measured_36)(xs), viewL(__dict_Monoid_35)(__dict_Measured_36)(ys));
                        if (_2544.value0 instanceof NilL && _2544.value1 instanceof NilL) {
                            return true;
                        };
                        if (_2544.value0 instanceof NilL) {
                            return false;
                        };
                        if (_2544.value1 instanceof NilL) {
                            return false;
                        };
                        if (_2544.value0 instanceof ConsL && _2544.value1 instanceof ConsL) {
                            var _2551 = Prelude["=="](__dict_Eq_37)(_2544.value0.value0)(_2544.value1.value0);
                            if (_2551) {
                                var ys$prime$prime = Data_Lazy.force(_2544.value1.value1);
                                var xs$prime$prime = Data_Lazy.force(_2544.value0.value1);
                                var __tco___dict_Monoid_35 = __dict_Monoid_35;
                                var __tco___dict_Measured_36 = __dict_Measured_36;
                                var __tco___dict_Eq_37 = __dict_Eq_37;
                                __dict_Monoid_35 = __tco___dict_Monoid_35;
                                __dict_Measured_36 = __tco___dict_Measured_36;
                                __dict_Eq_37 = __tco___dict_Eq_37;
                                xs = xs$prime$prime;
                                ys = ys$prime$prime;
                                continue tco;
                            };
                            if (!_2551) {
                                return false;
                            };
                            throw new Error("Failed pattern match");
                        };
                        throw new Error("Failed pattern match");
                    };
                };
            };
        };
    };
};
var head = function (__dict_Monoid_38) {
    return function (__dict_Measured_39) {
        return function (x) {
            var _2558 = viewL(__dict_Monoid_38)(__dict_Measured_39)(x);
            if (_2558 instanceof ConsL) {
                return new Data_Maybe.Just(_2558.value0);
            };
            if (_2558 instanceof NilL) {
                return Data_Maybe.Nothing.value;
            };
            throw new Error("Failed pattern match");
        };
    };
};
var isEmpty = function (__dict_Monoid_40) {
    return function (__dict_Measured_41) {
        return function (x) {
            var _2561 = viewL(__dict_Monoid_40)(__dict_Measured_41)(x);
            if (_2561 instanceof NilL) {
                return true;
            };
            if (_2561 instanceof ConsL) {
                return false;
            };
            throw new Error("Failed pattern match");
        };
    };
};
var tail = function (__dict_Monoid_42) {
    return function (__dict_Measured_43) {
        return function (x) {
            var _2564 = viewL(__dict_Monoid_42)(__dict_Measured_43)(x);
            if (_2564 instanceof ConsL) {
                return new Data_Maybe.Just(Data_Lazy.force(_2564.value1));
            };
            if (_2564 instanceof NilL) {
                return Data_Maybe.Nothing.value;
            };
            throw new Error("Failed pattern match");
        };
    };
};
var unfoldLeft = function (__dict_Unfoldable_44) {
    return function (__dict_Monoid_45) {
        return function (__dict_Measured_46) {
            var step = function (tree) {
                var _2567 = viewL(__dict_Monoid_45)(__dict_Measured_46)(tree);
                if (_2567 instanceof ConsL) {
                    return new Data_Maybe.Just(new Data_Tuple.Tuple(_2567.value0, Data_Lazy.force(_2567.value1)));
                };
                if (_2567 instanceof NilL) {
                    return Data_Maybe.Nothing.value;
                };
                throw new Error("Failed pattern match");
            };
            return Data_Unfoldable.unfoldr(__dict_Unfoldable_44)(step);
        };
    };
};
var viewR = function (__dict_Monoid_47) {
    return function (__dict_Measured_48) {
        return function (_522) {
            if (_522 instanceof Empty) {
                return NilR.value;
            };
            if (_522 instanceof Single) {
                return new SnocR(lazyEmpty, _522.value0);
            };
            if (_522 instanceof Deep) {
                return new SnocR(Data_Lazy.defer(function (_508) {
                    return deepR(__dict_Monoid_47)(__dict_Measured_48)(_522.value1)(_522.value2)(initDigit(_522.value3));
                }), lastDigit(_522.value3));
            };
            throw new Error("Failed pattern match");
        };
    };
};
var deepR = function (__dict_Monoid_49) {
    return function (__dict_Measured_50) {
        return function (pr) {
            return function (m) {
                return function (_523) {
                    if (_523.length === 0) {
                        var _2580 = viewR(__dict_Monoid_49)(measuredNode)(Data_Lazy.force(m));
                        if (_2580 instanceof NilR) {
                            return toFingerTree(__dict_Monoid_49)(__dict_Measured_50)(Data_Foldable.foldableArray)(pr);
                        };
                        if (_2580 instanceof SnocR) {
                            return deep(__dict_Monoid_49)(__dict_Measured_50)(pr)(_2580.value0)(nodeToDigit(_2580.value1));
                        };
                        throw new Error("Failed pattern match");
                    };
                    return deep(__dict_Monoid_49)(__dict_Measured_50)(pr)(m)(_523);
                };
            };
        };
    };
};
var init = function (__dict_Monoid_51) {
    return function (__dict_Measured_52) {
        return function (x) {
            var _2583 = viewR(__dict_Monoid_51)(__dict_Measured_52)(x);
            if (_2583 instanceof SnocR) {
                return new Data_Maybe.Just(Data_Lazy.force(_2583.value0));
            };
            if (_2583 instanceof NilR) {
                return Data_Maybe.Nothing.value;
            };
            throw new Error("Failed pattern match");
        };
    };
};
var last = function (__dict_Monoid_53) {
    return function (__dict_Measured_54) {
        return function (x) {
            var _2586 = viewR(__dict_Monoid_53)(__dict_Measured_54)(x);
            if (_2586 instanceof SnocR) {
                return new Data_Maybe.Just(_2586.value1);
            };
            if (_2586 instanceof NilR) {
                return Data_Maybe.Nothing.value;
            };
            throw new Error("Failed pattern match");
        };
    };
};
var unfoldRight = function (__dict_Unfoldable_55) {
    return function (__dict_Monoid_56) {
        return function (__dict_Measured_57) {
            var step = function (tree) {
                var _2589 = viewR(__dict_Monoid_56)(__dict_Measured_57)(tree);
                if (_2589 instanceof SnocR) {
                    return new Data_Maybe.Just(new Data_Tuple.Tuple(_2589.value1, Data_Lazy.force(_2589.value0)));
                };
                if (_2589 instanceof NilR) {
                    return Data_Maybe.Nothing.value;
                };
                throw new Error("Failed pattern match");
            };
            return Data_Unfoldable.unfoldr(__dict_Unfoldable_55)(step);
        };
    };
};
var unsafeSplitTree = function (__dict_Monoid_58) {
    return function (__dict_Measured_59) {
        return function (p) {
            return function (i) {
                return function (_527) {
                    if (_527 instanceof Single) {
                        return new LazySplit(lazyEmpty, _527.value0, lazyEmpty);
                    };
                    if (_527 instanceof Deep) {
                        var vpr = Prelude["<>"](__dict_Monoid_58["__superclass_Prelude.Semigroup_0"]())(i)(Data_Sequence_Internal.measure(Data_Sequence_Internal.measuredArray(__dict_Monoid_58)(__dict_Measured_59))(_527.value1));
                        var _2596 = p(vpr);
                        if (_2596) {
                            var _2597 = splitDigit(__dict_Monoid_58)(__dict_Measured_59)(p)(i)(_527.value1);
                            return new LazySplit(Data_Lazy.defer(function (_509) {
                                return toFingerTree(__dict_Monoid_58)(__dict_Measured_59)(Data_Foldable.foldableArray)(_2597.value0);
                            }), _2597.value1, Data_Lazy.defer(function (_510) {
                                return deepL(__dict_Monoid_58)(__dict_Measured_59)(_2597.value2)(_527.value2)(_527.value3);
                            }));
                        };
                        if (!_2596) {
                            var vm = Prelude["<>"](__dict_Monoid_58["__superclass_Prelude.Semigroup_0"]())(vpr)(Data_Sequence_Internal.measure(Data_Sequence_Internal.measuredLazy(__dict_Monoid_58)(measuredFingerTree(__dict_Monoid_58)(measuredNode)))(_527.value2));
                            var _2603 = p(vm);
                            if (_2603) {
                                var _2604 = unsafeSplitTree(__dict_Monoid_58)(measuredNode)(p)(vpr)(Data_Lazy.force(_527.value2));
                                var _2605 = splitDigit(__dict_Monoid_58)(__dict_Measured_59)(p)(Prelude["<>"](__dict_Monoid_58["__superclass_Prelude.Semigroup_0"]())(vpr)(Data_Sequence_Internal.measure(Data_Sequence_Internal.measuredLazy(__dict_Monoid_58)(measuredFingerTree(__dict_Monoid_58)(measuredNode)))(_2604.value0)))(nodeToDigit(_2604.value1));
                                return new LazySplit(Data_Lazy.defer(function (_511) {
                                    return deepR(__dict_Monoid_58)(__dict_Measured_59)(_527.value1)(_2604.value0)(_2605.value0);
                                }), _2605.value1, Data_Lazy.defer(function (_512) {
                                    return deepL(__dict_Monoid_58)(__dict_Measured_59)(_2605.value2)(_2604.value2)(_527.value3);
                                }));
                            };
                            if (!_2603) {
                                var _2614 = splitDigit(__dict_Monoid_58)(__dict_Measured_59)(p)(vm)(_527.value3);
                                return new LazySplit(Data_Lazy.defer(function (_513) {
                                    return deepR(__dict_Monoid_58)(__dict_Measured_59)(_527.value1)(_527.value2)(_2614.value0);
                                }), _2614.value1, Data_Lazy.defer(function (_514) {
                                    return toFingerTree(__dict_Monoid_58)(__dict_Measured_59)(Data_Foldable.foldableArray)(_2614.value2);
                                }));
                            };
                            throw new Error("Failed pattern match");
                        };
                        throw new Error("Failed pattern match");
                    };
                    throw new Error("Failed pattern match");
                };
            };
        };
    };
};
var cons = function (__dict_Monoid_60) {
    return function (__dict_Measured_61) {
        return function (a) {
            return function (_518) {
                if (_518 instanceof Empty) {
                    return new Single(a);
                };
                if (_518 instanceof Single) {
                    return deep(__dict_Monoid_60)(__dict_Measured_61)([ a ])(lazyEmpty)([ _518.value0 ]);
                };
                if (_518 instanceof Deep && _518.value1.length === 4) {
                    var forcedM = Data_Lazy.force(_518.value2);
                    return deep(__dict_Monoid_60)(__dict_Measured_61)([ a, _518.value1[0] ])(Data_Lazy.defer(function (_505) {
                        return cons(__dict_Monoid_60)(measuredNode)(node3(__dict_Monoid_60)(__dict_Measured_61)(_518.value1[1])(_518.value1[2])(_518.value1[3]))(forcedM);
                    }))(_518.value3);
                };
                if (_518 instanceof Deep) {
                    return deep(__dict_Monoid_60)(__dict_Measured_61)(Prelude[":"](a)(_518.value1))(_518.value2)(_518.value3);
                };
                throw new Error("Failed pattern match");
            };
        };
    };
};
var consAll = function (__dict_Monoid_62) {
    return function (__dict_Measured_63) {
        return function (__dict_Foldable_64) {
            return Prelude.flip(Data_Foldable.foldr(__dict_Foldable_64)(cons(__dict_Monoid_62)(__dict_Measured_63)));
        };
    };
};
var filter = function (__dict_Monoid_65) {
    return function (__dict_Measured_66) {
        return function (p) {
            return Data_Foldable.foldr(foldableFingerTree)(function (x) {
                return function (acc) {
                    var _2640 = p(x);
                    if (_2640) {
                        return cons(__dict_Monoid_65)(__dict_Measured_66)(x)(acc);
                    };
                    if (!_2640) {
                        return acc;
                    };
                    throw new Error("Failed pattern match");
                };
            })(Empty.value);
        };
    };
};
var split = function (__dict_Monoid_67) {
    return function (__dict_Measured_68) {
        return function (p) {
            return function (_528) {
                if (_528 instanceof Empty) {
                    return new Data_Tuple.Tuple(lazyEmpty, lazyEmpty);
                };
                var _2643 = p(Data_Sequence_Internal.measure(measuredFingerTree(__dict_Monoid_67)(__dict_Measured_68))(_528));
                if (_2643) {
                    var _2644 = unsafeSplitTree(__dict_Monoid_67)(__dict_Measured_68)(p)(Data_Monoid.mempty(__dict_Monoid_67))(_528);
                    return new Data_Tuple.Tuple(_2644.value0, Data_Lazy.defer(function (_515) {
                        return cons(__dict_Monoid_67)(__dict_Measured_68)(_2644.value1)(Data_Lazy.force(_2644.value2));
                    }));
                };
                if (!_2643) {
                    return new Data_Tuple.Tuple(Data_Lazy.defer(function (_516) {
                        return _528;
                    }), lazyEmpty);
                };
                throw new Error("Failed pattern match");
            };
        };
    };
};
var compareFingerTree = function (__copy___dict_Monoid_69) {
    return function (__copy___dict_Measured_70) {
        return function (__copy___dict_Ord_71) {
            return function (__copy_xs) {
                return function (__copy_ys) {
                    var __dict_Monoid_69 = __copy___dict_Monoid_69;
                    var __dict_Measured_70 = __copy___dict_Measured_70;
                    var __dict_Ord_71 = __copy___dict_Ord_71;
                    var xs = __copy_xs;
                    var ys = __copy_ys;
                    tco: while (true) {
                        var _2650 = new Data_Tuple.Tuple(viewL(__dict_Monoid_69)(__dict_Measured_70)(xs), viewL(__dict_Monoid_69)(__dict_Measured_70)(ys));
                        if (_2650.value0 instanceof NilL && _2650.value1 instanceof NilL) {
                            return Prelude.EQ.value;
                        };
                        if (_2650.value0 instanceof NilL) {
                            return Prelude.LT.value;
                        };
                        if (_2650.value1 instanceof NilL) {
                            return Prelude.GT.value;
                        };
                        if (_2650.value0 instanceof ConsL && _2650.value1 instanceof ConsL) {
                            var _2657 = Prelude.compare(__dict_Ord_71)(_2650.value0.value0)(_2650.value1.value0);
                            if (_2657 instanceof Prelude.EQ) {
                                var ys$prime$prime = Data_Lazy.force(_2650.value1.value1);
                                var xs$prime$prime = Data_Lazy.force(_2650.value0.value1);
                                var __tco___dict_Monoid_69 = __dict_Monoid_69;
                                var __tco___dict_Measured_70 = __dict_Measured_70;
                                var __tco___dict_Ord_71 = __dict_Ord_71;
                                __dict_Monoid_69 = __tco___dict_Monoid_69;
                                __dict_Measured_70 = __tco___dict_Measured_70;
                                __dict_Ord_71 = __tco___dict_Ord_71;
                                xs = xs$prime$prime;
                                ys = ys$prime$prime;
                                continue tco;
                            };
                            return _2657;
                        };
                        throw new Error("Failed pattern match");
                    };
                };
            };
        };
    };
};
var app3 = function (__dict_Monoid_72) {
    return function (__dict_Measured_73) {
        return function (_524) {
            return function (ts) {
                return function (_525) {
                    if (_524 instanceof Empty) {
                        return consAll(__dict_Monoid_72)(__dict_Measured_73)(Data_Foldable.foldableArray)(ts)(_525);
                    };
                    if (_525 instanceof Empty) {
                        return snocAll(__dict_Monoid_72)(__dict_Measured_73)(Data_Foldable.foldableArray)(_524)(ts);
                    };
                    if (_524 instanceof Single) {
                        return cons(__dict_Monoid_72)(__dict_Measured_73)(_524.value0)(consAll(__dict_Monoid_72)(__dict_Measured_73)(Data_Foldable.foldableArray)(ts)(_525));
                    };
                    if (_525 instanceof Single) {
                        return snoc(__dict_Monoid_72)(__dict_Measured_73)(snocAll(__dict_Monoid_72)(__dict_Measured_73)(Data_Foldable.foldableArray)(_524)(ts))(_525.value0);
                    };
                    if (_524 instanceof Deep && _525 instanceof Deep) {
                        var computeM$prime = function (_542) {
                            return app3(__dict_Monoid_72)(measuredNode)(Data_Lazy.force(_524.value2))(nodes(__dict_Monoid_72)(__dict_Measured_73)(Prelude["<>"](Data_Array.semigroupArray)(_524.value3)(Prelude["<>"](Data_Array.semigroupArray)(ts)(_525.value1))))(Data_Lazy.force(_525.value2));
                        };
                        return deep(__dict_Monoid_72)(__dict_Measured_73)(_524.value1)(Data_Lazy.defer(computeM$prime))(_525.value3);
                    };
                    throw new Error("Failed pattern match");
                };
            };
        };
    };
};
var append = function (__dict_Monoid_74) {
    return function (__dict_Measured_75) {
        return function (xs) {
            return function (ys) {
                return app3(__dict_Monoid_74)(__dict_Measured_75)(xs)([  ])(ys);
            };
        };
    };
};
var semigroupFingerTree = function (__dict_Monoid_76) {
    return function (__dict_Measured_77) {
        return new Prelude.Semigroup(append(__dict_Monoid_76)(__dict_Measured_77));
    };
};
module.exports = {
    LazySplit: LazySplit, 
    Split: Split, 
    NilR: NilR, 
    SnocR: SnocR, 
    NilL: NilL, 
    ConsL: ConsL, 
    Empty: Empty, 
    Single: Single, 
    Deep: Deep, 
    Node2: Node2, 
    Node3: Node3, 
    unfoldRight: unfoldRight, 
    unfoldLeft: unfoldLeft, 
    filter: filter, 
    split: split, 
    unsafeSplitTree: unsafeSplitTree, 
    splitDigit: splitDigit, 
    append: append, 
    nodes: nodes, 
    app3: app3, 
    init: init, 
    last: last, 
    deepR: deepR, 
    viewR: viewR, 
    initDigit: initDigit, 
    lastDigit: lastDigit, 
    tail: tail, 
    head: head, 
    isEmpty: isEmpty, 
    deepL: deepL, 
    viewL: viewL, 
    tailDigit: tailDigit, 
    headDigit: headDigit, 
    toFingerTree: toFingerTree, 
    snocAll: snocAll, 
    consAll: consAll, 
    snoc: snoc, 
    cons: cons, 
    compareFingerTree: compareFingerTree, 
    eqFingerTree: eqFingerTree, 
    deep: deep, 
    lazyEmpty: lazyEmpty, 
    nodeToDigit: nodeToDigit, 
    node3: node3, 
    node2: node2, 
    showNode: showNode, 
    functorNode: functorNode, 
    foldableNode: foldableNode, 
    traversableNode: traversableNode, 
    measuredNode: measuredNode, 
    showFingerTree: showFingerTree, 
    semigroupFingerTree: semigroupFingerTree, 
    functorFingerTree: functorFingerTree, 
    foldableFingerTree: foldableFingerTree, 
    traversableFingerTree: traversableFingerTree, 
    measuredFingerTree: measuredFingerTree, 
    functorViewL: functorViewL
};
