// Control.Monad.Eff.Class/index.js
// Generated by psc-make version 0.6.9.5
"use strict";
var Prelude = require("Prelude");
var Control_Monad_Trans = require("Control.Monad.Trans");
var Data_Monoid = require("Data.Monoid");
var Control_Monad_Eff = require("Control.Monad.Eff");
var Control_Monad_Maybe_Trans = require("Control.Monad.Maybe.Trans");
var Control_Monad_Error_Trans = require("Control.Monad.Error.Trans");
var Control_Monad_State_Trans = require("Control.Monad.State.Trans");
var Control_Monad_Writer_Trans = require("Control.Monad.Writer.Trans");
var Control_Monad_Reader_Trans = require("Control.Monad.Reader.Trans");
var Control_Monad_RWS_Trans = require("Control.Monad.RWS.Trans");
var MonadEff = function (__superclass_Prelude$dotMonad_0, liftEff) {
    this["__superclass_Prelude.Monad_0"] = __superclass_Prelude$dotMonad_0;
    this.liftEff = liftEff;
};
var monadEffEff = new MonadEff(function () {
    return Control_Monad_Eff.monadEff;
}, Prelude.id(Prelude.categoryArr));
var liftEff = function (dict) {
    return dict.liftEff;
};
var monadEffError = function (__dict_Monad_0) {
    return function (__dict_MonadEff_1) {
        return new MonadEff(function () {
            return Control_Monad_Error_Trans.monadErrorT(__dict_Monad_0);
        }, Prelude["<<<"](Prelude.semigroupoidArr)(Control_Monad_Trans.lift(Control_Monad_Error_Trans.monadTransErrorT)(__dict_Monad_0))(liftEff(__dict_MonadEff_1)));
    };
};
var monadEffMaybe = function (__dict_Monad_2) {
    return function (__dict_MonadEff_3) {
        return new MonadEff(function () {
            return Control_Monad_Maybe_Trans.monadMaybeT(__dict_Monad_2);
        }, Prelude["<<<"](Prelude.semigroupoidArr)(Control_Monad_Trans.lift(Control_Monad_Maybe_Trans.monadTransMaybeT)(__dict_Monad_2))(liftEff(__dict_MonadEff_3)));
    };
};
var monadEffRWS = function (__dict_Monad_4) {
    return function (__dict_Monoid_5) {
        return function (__dict_MonadEff_6) {
            return new MonadEff(function () {
                return Control_Monad_RWS_Trans.monadRWST(__dict_Monad_4)(__dict_Monoid_5);
            }, Prelude["<<<"](Prelude.semigroupoidArr)(Control_Monad_Trans.lift(Control_Monad_RWS_Trans.monadTransRWST(__dict_Monoid_5))(__dict_Monad_4))(liftEff(__dict_MonadEff_6)));
        };
    };
};
var monadEffReader = function (__dict_Monad_7) {
    return function (__dict_MonadEff_8) {
        return new MonadEff(function () {
            return Control_Monad_Reader_Trans.monadReaderT(__dict_Monad_7);
        }, Prelude["<<<"](Prelude.semigroupoidArr)(Control_Monad_Trans.lift(Control_Monad_Reader_Trans.monadTransReaderT)(__dict_Monad_7))(liftEff(__dict_MonadEff_8)));
    };
};
var monadEffState = function (__dict_Monad_9) {
    return function (__dict_MonadEff_10) {
        return new MonadEff(function () {
            return Control_Monad_State_Trans.monadStateT(__dict_Monad_9);
        }, Prelude["<<<"](Prelude.semigroupoidArr)(Control_Monad_Trans.lift(Control_Monad_State_Trans.monadTransStateT)(__dict_Monad_9))(liftEff(__dict_MonadEff_10)));
    };
};
var monadEffWriter = function (__dict_Monad_11) {
    return function (__dict_Monoid_12) {
        return function (__dict_MonadEff_13) {
            return new MonadEff(function () {
                return Control_Monad_Writer_Trans.monadWriterT(__dict_Monoid_12)(__dict_Monad_11);
            }, Prelude["<<<"](Prelude.semigroupoidArr)(Control_Monad_Trans.lift(Control_Monad_Writer_Trans.monadTransWriterT(__dict_Monoid_12))(__dict_Monad_11))(liftEff(__dict_MonadEff_13)));
        };
    };
};
module.exports = {
    MonadEff: MonadEff, 
    liftEff: liftEff, 
    monadEffEff: monadEffEff, 
    monadEffMaybe: monadEffMaybe, 
    monadEffError: monadEffError, 
    monadEffState: monadEffState, 
    monadEffWriter: monadEffWriter, 
    monadEffReader: monadEffReader, 
    monadEffRWS: monadEffRWS
};
