// Test.Spec.Reporter/index.js
// Generated by psc-make version 0.6.9.5
"use strict";
var Prelude = require("Prelude");
var Test_Spec_Errors = require("Test.Spec.Errors");
var Data_String = require("Data.String");
var Data_Foldable = require("Data.Foldable");
var Data_Array = require("Data.Array");
var Debug_Trace = require("Debug.Trace");
var Control_Monad_Eff = require("Control.Monad.Eff");
var Control_Monad_Eff_Exception = require("Control.Monad.Eff.Exception");
var Control_Monad_Extras = require("Control.Monad.Extras");
var Test_Spec = require("Test.Spec");
var Test_Spec_Console = require("Test.Spec.Console");
var Test_Spec_Summary = require("Test.Spec.Summary");
var Describe = (function () {
    function Describe(value0) {
        this.value0 = value0;
    };
    Describe.create = function (value0) {
        return new Describe(value0);
    };
    return Describe;
})();
var It = (function () {
    function It(value0, value1) {
        this.value0 = value0;
        this.value1 = value1;
    };
    It.create = function (value0) {
        return function (value1) {
            return new It(value0, value1);
        };
    };
    return It;
})();
var Pending = (function () {
    function Pending(value0) {
        this.value0 = value0;
    };
    Pending.create = function (value0) {
        return new Pending(value0);
    };
    return Pending;
})();
var showEntry = new Prelude.Show(function (_642) {
    if (_642 instanceof Describe) {
        return "Describe \"" + (Data_String.joinWith(" » ")(_642.value0) + "\"");
    };
    if (_642 instanceof It && _642.value1 instanceof Test_Spec.Success) {
        return "It \"" + (_642.value0 + "\" Success");
    };
    if (_642 instanceof It && _642.value1 instanceof Test_Spec.Failure) {
        return "It \"" + (_642.value0 + ("\" (Failure \"" + (Test_Spec_Errors.errorMessage(_642.value1.value0) + "\")")));
    };
    if (_642 instanceof Pending) {
        return "Pending \"" + (_642.value0 + "\"");
    };
    throw new Error("Failed pattern match");
});
var eqEntry = new Prelude.Eq(function (e1) {
    return function (e2) {
        return !Prelude["=="](eqEntry)(e1)(e2);
    };
}, function (_640) {
    return function (_641) {
        if (_640 instanceof Describe && _641 instanceof Describe) {
            return Prelude["=="](Prelude.eqArray(Prelude.eqString))(_640.value0)(_641.value0);
        };
        if (_640 instanceof It && (_640.value1 instanceof Test_Spec.Success && (_641 instanceof It && _641.value1 instanceof Test_Spec.Success))) {
            return _640.value0 === _641.value0;
        };
        if (_640 instanceof It && (_640.value1 instanceof Test_Spec.Failure && (_641 instanceof It && _641.value1 instanceof Test_Spec.Failure))) {
            return _640.value0 === _641.value0 && Test_Spec_Errors.errorMessage(_640.value1.value0) === Test_Spec_Errors.errorMessage(_641.value1.value0);
        };
        if (_640 instanceof It && _641 instanceof It) {
            return false;
        };
        if (_640 instanceof Pending && _641 instanceof Pending) {
            return _640.value0 === _641.value0;
        };
        throw new Error("Failed pattern match");
    };
});
var countDescribes = function (groups) {
    var f = function (c) {
        return function (_643) {
            if (_643 instanceof Describe) {
                return c + 1;
            };
            return c;
        };
    };
    return Data_Foldable.foldl(Data_Foldable.foldableArray)(f)(0)(groups);
};
var collapse = function (_639) {
    if (_639 instanceof Test_Spec.It) {
        return [ new It(_639.value0, _639.value1) ];
    };
    if (_639 instanceof Test_Spec.Pending) {
        return [ new Pending(_639.value0) ];
    };
    if (_639 instanceof Test_Spec.Describe) {
        var sub = Data_Array.concatMap(collapse)(_639.value1);
        var prependName = function (_644) {
            if (_644 instanceof Describe) {
                return new Describe(Prelude[":"](_639.value0)(_644.value0));
            };
            return _644;
        };
        var c = countDescribes(sub);
        var _3047 = c === 0;
        if (_3047) {
            return Prelude[":"](new Describe([ _639.value0 ]))(sub);
        };
        if (!_3047) {
            return Data_Array.map(prependName)(sub);
        };
        throw new Error("Failed pattern match");
    };
    throw new Error("Failed pattern match");
};
module.exports = {
    Describe: Describe, 
    It: It, 
    Pending: Pending, 
    collapse: collapse, 
    eqEntry: eqEntry, 
    showEntry: showEntry
};
