// Test.Unit.Console/index.js
// Generated by psc-make version 0.6.9.5
"use strict";
var Prelude = require("Prelude");
var Control_Monad_Eff = require("Control.Monad.Eff");

  var hasStderr;
  try { hasStderr = !!process.stderr; } catch (e) { hasStderr = false; }
;

  function consoleLog(s) {
    return function() {
      console.log(s);
    };
  };

  function consoleError(s) {
    return function() {
      console.error(s);
    };
  };

  function savePos() {
    process.stderr.write("\x1b[s");
  };

  function restorePos() {
    process.stderr.write("\x1b[u");
  };

  function eraseLine() {
    process.stderr.write("\x1b[K");
  };

  function print(s) {
    return function() {
      process.stderr.write("\x1b[33m" + s + "\x1b[0m");
    };
  };

  function printLabel(s) {
    return function() {
      process.stderr.write("\x1b[33;1m" + s + "\x1b[0m");
    };
  };

  function printFail(s) {
    return function() {
      process.stderr.write("\x1b[31;1m" + s + "\x1b[0m");
    };
  };

  function printPass(s) {
    return function() {
      process.stderr.write("\x1b[32m" + s + "\x1b[0m");
    };
  };
module.exports = {
    printPass: printPass, 
    printFail: printFail, 
    printLabel: printLabel, 
    print: print, 
    eraseLine: eraseLine, 
    restorePos: restorePos, 
    savePos: savePos, 
    consoleError: consoleError, 
    consoleLog: consoleLog, 
    hasStderr: hasStderr
};
