// Data.Monoid.Any/index.js
// Generated by psc-make version 0.6.9.5
"use strict";
var Prelude = require("Prelude");
var Data_Monoid = require("Data.Monoid");
var Any = function (x) {
    return x;
};
var showAny = new Prelude.Show(function (_201) {
    return "Any (" + (Prelude.show(Prelude.showBoolean)(_201) + ")");
});
var semigroupAny = new Prelude.Semigroup(function (_202) {
    return function (_203) {
        return _202 || _203;
    };
});
var runAny = function (_194) {
    return _194;
};
var monoidAny = new Data_Monoid.Monoid(function () {
    return semigroupAny;
}, false);
var eqAny = new Prelude.Eq(function (_197) {
    return function (_198) {
        return _197 !== _198;
    };
}, function (_195) {
    return function (_196) {
        return _195 === _196;
    };
});
var ordAny = new Prelude.Ord(function () {
    return eqAny;
}, function (_199) {
    return function (_200) {
        return Prelude.compare(Prelude.ordBoolean)(_199)(_200);
    };
});
module.exports = {
    Any: Any, 
    runAny: runAny, 
    eqAny: eqAny, 
    ordAny: ordAny, 
    showAny: showAny, 
    semigroupAny: semigroupAny, 
    monoidAny: monoidAny
};
