// Data.Monoid.Any/index.js
// Generated by psc-make version 0.6.9.5
"use strict";
var Prelude = require("Prelude");
var Data_Monoid = require("Data.Monoid");
var Any = function (x) {
    return x;
};
var showAny = new Prelude.Show(function (_192) {
    return "Any (" + (Prelude.show(Prelude.showBoolean)(_192) + ")");
});
var semigroupAny = new Prelude.Semigroup(function (_193) {
    return function (_194) {
        return _193 || _194;
    };
});
var runAny = function (_185) {
    return _185;
};
var monoidAny = new Data_Monoid.Monoid(function () {
    return semigroupAny;
}, false);
var eqAny = new Prelude.Eq(function (_188) {
    return function (_189) {
        return _188 !== _189;
    };
}, function (_186) {
    return function (_187) {
        return _186 === _187;
    };
});
var ordAny = new Prelude.Ord(function () {
    return eqAny;
}, function (_190) {
    return function (_191) {
        return Prelude.compare(Prelude.ordBoolean)(_190)(_191);
    };
});
module.exports = {
    Any: Any, 
    runAny: runAny, 
    eqAny: eqAny, 
    ordAny: ordAny, 
    showAny: showAny, 
    semigroupAny: semigroupAny, 
    monoidAny: monoidAny
};
