// Data.Monoid.Any/index.js
// Generated by psc-make version 0.6.9.5
"use strict";
var Prelude = require("Prelude");
var Data_Monoid = require("Data.Monoid");
var Any = function (x) {
    return x;
};
var showAny = new Prelude.Show(function (_145) {
    return "Any (" + (Prelude.show(Prelude.showBoolean)(_145) + ")");
});
var semigroupAny = new Prelude.Semigroup(function (_146) {
    return function (_147) {
        return _146 || _147;
    };
});
var runAny = function (_138) {
    return _138;
};
var monoidAny = new Data_Monoid.Monoid(function () {
    return semigroupAny;
}, false);
var eqAny = new Prelude.Eq(function (_141) {
    return function (_142) {
        return _141 !== _142;
    };
}, function (_139) {
    return function (_140) {
        return _139 === _140;
    };
});
var ordAny = new Prelude.Ord(function () {
    return eqAny;
}, function (_143) {
    return function (_144) {
        return Prelude.compare(Prelude.ordBoolean)(_143)(_144);
    };
});
module.exports = {
    Any: Any, 
    runAny: runAny, 
    eqAny: eqAny, 
    ordAny: ordAny, 
    showAny: showAny, 
    semigroupAny: semigroupAny, 
    monoidAny: monoidAny
};
